---
title: "GRCCA paper figures for poster"
author: "Rachel Smith"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width = 2.5, fig.height = 2, collapse = TRUE, warning = FALSE, message = FALSE, echo = FALSE)

## Set paths and load data
base_dir <- "~/Documents/PhD/projects/sgacc_wgcna_grcca/"
poster_figures_dir <- paste0(base_dir, "outputs/figures/poster/")
tables_dir <- paste0(base_dir, "outputs/tables/")
analysis_objects_dir <- paste0(base_dir, "outputs/objects/")

source(paste0(base_dir, "scripts/full_analysis_scripts/scripts/load_WGCNA_res.R"))

## Load data objects for plots
directories <- list.files(analysis_objects_dir)
for (dir in directories) {
    print(dir)
    objects <- list.files(paste0(analysis_objects_dir, dir))
    #objects <- objects[!str_detect(objects, "null|archive")]
    for (obj in objects){
        print(paste0("loading ", obj, "...."))
        load(paste0(analysis_objects_dir, dir, "/", obj))
    }
}
```

# Fig 1

## A | Covariate (Y) structure correlations

```{r Fig1a.plot}
fig1a <- df_y_res %>% 
    mutate(significant = ifelse(abs(z_score) >= 2 & p_adj < 0.05, "*", "")) %>% 
    
    ggplot(aes(x = pearsons_r, y = reorder(covariate, pearsons_r))) +
    geom_point(aes(size = -log10(p_adj), fill = pearsons_r), shape = 21) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    geom_text(aes(label = significant), size = 6, vjust = 0.75, hjust = -1.5, color = "black") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1, 1), guide = "none") +
    scale_size_continuous(range = c(2, 6)) +
    guides(size = guide_legend(title = "-log10(FDR)")) +
    coord_cartesian(clip = "off") +
    xlim(c(-1, 1)) +
    labs(y = NULL, x = expression("Structure correlation (" * italic(r)[y] * ")"), 
         title = "A | Covariate structure correlations") +
    theme(legend.position = c(0.075, 0.6), 
          legend.box = "horizontal",
          #legend.justification = "center",
          legend.key.size = unit(0.1, 'cm')
          #legend.key.height = unit(0.29, "cm"),
          #legend.key.width = unit(0.30, "cm")
    )
```

```{r Fig1a.save, fig.width = 2.5, fig.height = 2}
fig1a
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "1A.covariate_structCor", .x), width = fig_width, height = fig_height)
)
```

## B | Top 10 genes (X) by structure correlation

```{r Fig1b.prep}
df_fig1b <- df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) >= 2) %>%
    top_n(n = 10, wt = abs(pearsons_r)) %>%
    mutate(direction = ifelse(pearsons_r < 0, "2neg", "1pos"))
```

```{r Fig1b.plot}
fig1b <- df_fig1b %>% 
    ggplot(aes(x = pearsons_r, y = reorder(gene_symbol, pearsons_r))) +
    geom_point(aes(size = -log10(p_adj), fill = pearsons_r), shape = 21) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    facet_wrap(vars(direction), ncol = 1, scales = "free_y") +
    force_panelsizes(rows = c(16, 4)) +
    scale_fill_gradientn(colors = gene_weight_color_scale, 
                         limits = c(-0.57, 0.57), guide = "none") +
    scale_size_continuous(range = c(1, 4)) +
    guides(size = guide_legend(title = "-log10(FDR)")) +
    xlim(c(-1, 1)) +
    coord_cartesian(clip = "off") +
    labs(y = NULL, x = expression("Structure correlation (" * italic(r)[x] * ")"), 
         title = "B | Top 10 genes by \nstructure correlation"
    ) +
    theme(strip.text = element_blank(), 
          panel.spacing = unit(0.5, "lines"),
          legend.position = c(0.05, 0.7), 
          legend.box = "horizontal",
          legend.key.size = unit(0.35, "cm")
    )
```

```{r Fig1b.save, fig.height = 2, fig.width = 2.5}
fig1b
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "1B.gene_structCor", .x), width = fig_width, height = fig_height)
)
```

## C | Expression directions of top positively- and negatively-weighted genes

```{r Fig1c.prep}
top_gene <- df_x_res %>% arrange(-abs(pearsons_r)) %>% head(1) %>% pull(ensembl_gene_id) # BCL7A
df_fig1c <- df_vsd_regress %>% 
    dplyr::select(sample, all_of(c(top_gene))) %>% 
    pivot_longer(contains("ENSG"), names_to = "ensembl_gene_id", values_to = "expr") %>% 
    left_join(df_ensembl_to_symbol) %>% 
    mutate(dx = case_when(
        str_detect(sample, "control") ~ "Control",
        str_detect(sample, "bipolar") ~ "BD",
        str_detect(sample, "mdd") ~ "MDD",
        str_detect(sample, "schizo") ~ "SCZ"
    ) %>% 
        factor(levels = names(dx_colors)))
```

```{r Fig1c.plot}
fig1c <- df_fig1c %>% 
    ggplot(aes(x = dx, y = expr)) +
    geom_point(aes(color = dx), position = position_jitter(width = 0.2), size = 0.5) +
    geom_boxplot(aes(color = dx, fill = dx), alpha = 0.1, outlier.shape = NA) +
    geom_hline(yintercept = 0, color = "darkgrey") +
    facet_wrap(vars(gene_symbol), scales = "free_y", nrow = 1) +
    #force_panelsizes(cols = c(1, 1)) +
    scale_fill_manual(values = dx_colors, guide = "none") +
    scale_color_manual(values = dx_colors, guide = "none") +
    labs(x = NULL, y = "Corrected expression value",
         title = "C | Gene expression values") +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9))
```

```{r Fig1c.save, fig.width = 2, fig.height = 2}
fig1c
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "1C.top_gene_expression_values", .x), width = 2, height = fig_height)
)
```

## D | GRCCA structure correlations (r) as they relate to expression values

```{r Fig1d.prep}
df_fig1d <- df_vsd_regress %>% 
    pivot_longer(contains("ENSG"), names_to = "ensembl_gene_id", values_to = "expression_value") %>% 
    mutate(dx = case_when(
        str_detect(sample, "control") ~ "Control",
        str_detect(sample, "bipolar") ~ "BD",
        str_detect(sample, "mdd") ~ "MDD",
        str_detect(sample, "schizo") ~ "SCZ"
    ) %>% factor(levels = names(dx_colors))) %>% 
    group_by(ensembl_gene_id, dx) %>% 
    summarise(mean_expr = mean(expression_value)) %>% 
    left_join(df_x_res)
```

```{r Fig1d.plot}
fig1d <- df_fig1d %>% 
    ggplot(aes(x = mean_expr, y = pearsons_r)) +
    geom_hline(yintercept = 0, color = "gray") +
    geom_vline(xintercept = 0, color = "gray") +
    geom_point(color = "black", size = 0.25) +
    geom_smooth(aes(color = dx), method = "lm", linewidth = 0.75) +
    stat_cor(aes(label = after_stat(r.label), color = dx), label.sep = "\n", size = 3,
             label.y.npc = "top", hjust = -0.1, vjust = -1.75, face = "bold") +
    facet_wrap(vars(dx), nrow = 1) +
    scale_color_manual(values = dx_colors) +
    scale_x_continuous(breaks = c(-0.4, 0, 0.4)) +
    coord_cartesian(clip = "off") +
    labs(x = "Mean expression (corrected)", y = expression("Structure correlation (" * italic(r)[x] * ")"),
         title = "D | Relationship between gene structure correlation & expression"
    ) +
    theme(legend.position = "none")
```

```{r Fig1d.save, fig.width = 5, fig.height = 2}
fig1d
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "1D.gene_strucCor_vs_Expr", .x), width = 5, height = fig_height)
)
```

# Fig 2

## A | GRCCA results are significantly and specifically enriched for SCZ risk genes

```{r Fig2a.prep}
df_fig2a <- df_grcca_fgsea_benchmark %>% 
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), "SCZ", "other") %>% factor(levels = c("SCZ", "other"))) %>% 
    mutate(benchmark_list = str_remove_all(benchmark_list, "SCZ|\\)|\\(") %>% 
               str_trim %>% 
               str_replace(", ", ",\n")
    ) %>% 
    mutate(label_color = ifelse(padj < 0.05, "white", "black")) %>% 
    mutate(box_color = ifelse(padj < 0.05, "black", "transparent")) %>% 
    mutate(label = round(NES, 2) %>% format(digits = 3))
    #mutate(sig = ifelse(padj < 0.05, scientific(padj, digits = 3), round(padj, 2) %>% format(digits = 3)))
   # mutate(sig = case_when(padj < 0.001 ~ "***", padj < 0.01 ~ "**", padj < 0.05 ~ "*", TRUE ~ ""))

## Identify genes to label in bottom panel
top_genes <- df_x_res %>% 
    dplyr::filter(ensembl_gene_id %in% benchmarking_lists[["SCZ (common, broad)"]]) %>% 
    top_n(n = 6, wt = abs(pearsons_r)) %>% 
    pull(ensembl_gene_id)

## Subset X res for genes in SCZ common broad list for bottom panel
df_fig2a_bottom <- df_x_res %>% 
    dplyr::filter(ensembl_gene_id %in% benchmarking_lists[["SCZ (common, broad)"]]) %>% 
    mutate(significant = ifelse(p_adj < 0.05, 1, 0) %>% factor) %>% 
    arrange(significant, abs(pearsons_r)) %>% 
    mutate(label = ifelse(ensembl_gene_id %in% top_genes, gene_symbol, ""))
```

```{r Fig2a.plot.top}
fig2a <- df_fig2a %>% 
    ggplot(aes(x = benchmark_list, y = 1)) +
    geom_tile(aes(fill = NES, color = I(box_color), alpha = abs(NES)), width = 0.95, height = 0.95, linewidth = 0.75) +
    geom_text(aes(label = label, color = I(label_color)), size = 3) +
    facet_wrap(vars(class), scales = "free", nrow = 2) +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1.58, 1.58)) +
    scale_alpha_continuous(guide = "none", limits = c(0.83, 1.58)) +
    guides(fill = guide_colorbar(title.hjust = 0.5, title.position = "top")) +
    labs(x = NULL, y = NULL,
         title = "A | Risk gene enrichment") +
    theme(axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_blank(),
          axis.text.x = element_text(margin = margin(t = -3)),
          strip.background = element_rect(fill = "white", color = "black"),
          legend.margin = margin(r = -15),
          legend.position = "bottom",
          legend.justification = "center",
          #legend.title = element_text(angle = 90),
          legend.key.width = unit(0.60, "cm"),
          legend.key.height = unit(0.15, "cm")
    )
```

```{r Fig2a.top.save, fig.height = 3, fig.width = 2.25}
fig2a
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "2A.Risk_gene_enrichment_grid", .x), width = 2.25, height = 3)
)
```

```{r Fig2a.plot.right}
scale_max <- df_x_res %>% pull(pearsons_r) %>% abs %>% max
jitter_pos <- position_jitter(width = 0.1, seed = 123)
fig2a_right <- df_x_res %>%
    ggplot(aes(x = pearsons_r)) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    geom_density() +
    geom_point(data = df_fig2a_bottom, 
               mapping = aes(x = pearsons_r, y = -1, fill = pearsons_r, color = significant), 
               pos = jitter_pos, shape = 21, size = 0.75
    ) +
    geom_text_repel(data = df_fig2a_bottom,
                    mapping = aes(x = pearsons_r, y = -1, label = label),
                    pos = jitter_pos, 
                    size = 3, box.padding = 0.5, max.overlaps = 50, force = 10, min.segment.length = 0
    ) +
    
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-scale_max, scale_max), guide = "none") +
    scale_color_manual(values = c("transparent", "black"), guide = "none") +
    scale_x_continuous(limits = c(-0.6, 0.6), breaks = c(-0.5, 0, 0.5)) +
    labs(x = expression(italic(r)[x]), y = NULL, 
         title = NULL) +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          panel.border = element_rect(fill = "transparent", color = "black", size = 1),
          axis.line = element_blank()
    )
```

```{r Fig2a.right.save, fig.height = 1.5, fig.width = 2}
fig2a_right
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "2A.Risk_gene_enrichment_distribution", .x), width = 2, height = 1.5)
)
```


## B | GRCCA cell-type enrichments

```{r Fig2b.plot}
fig2b <- df_grcca_fgsea_cell %>% 
    mutate(sig = case_when(padj < 0.001 ~ "***", padj < 0.01 ~ "**", padj < 0.05 ~ "*", TRUE ~ "")) %>% 
    
    ggplot(aes(x = NES, y = reorder(cell_type, NES))) +
    geom_col(aes(fill = NES), color = "gray", shape = 21) +
    geom_text(aes(label = sig, x = NES), vjust = 0.75, size = 4, check_overlap = TRUE) +
    scale_fill_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-3.39, 3.39)) +
    coord_cartesian(clip = "off") +
    labs(x = "NES", y = NULL,
         title = "B | Cell-type enrichment") +
    theme(legend.position = "none")
```

```{r Fig2b.save, fig.height = 1.5, fig.width = 2}
fig2b
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "2B.cell_type_enrichment", .x), width = 2, height = 1.5)
)
```

## C | Developmental expression trajectories of GRCCA deciles

```{r Fig2c.prep}
## Calculate the mean expression for each decile of GRCCA results across development
df_fig2c <- df_psychencode_grcca %>% #dplyr::filter(regioncode == "MFC") %>% 
    group_by(decile, window) %>% 
    summarise(median_expression = median(dev_expression),
              sd_expression = sd(dev_expression)
    )

## Rename windows with time period
psychencode_windows <- c("5-9",
                "12-13",
                "16-18",
                "19-22",
                "PCW35-PY0.3",
                "0.5-2.5",
                "2.8-10.7",
                "13-19",
                "21-64")
names(psychencode_windows) = 1:9  

## Define colors for deciles
decile_colors <- colorRampPalette(c(gene_weight_color_scale[5], "gray", gene_weight_color_scale[95]))(10) %>% rev
```

```{r Fig2c.plot}
fig2c <- df_fig2c %>% 
    ggplot(aes(x = window, y = median_expression, color = decile)) +
    geom_vline(xintercept = 5, color = "gray") +
    geom_smooth(method = "loess", se = FALSE) +
    scale_x_continuous(breaks = seq(1, length(psychencode_windows), 1),
                       labels = psychencode_windows) +
    scale_color_manual(values = decile_colors) +
    guides(color = guide_legend()) +
    labs(x = "← PCW | PY →", y = "Median expression",
         title = "C | Developmental expression") +
    theme(
        legend.key.size = unit(0.3, "cm"),
        axis.text.x = element_text(angle = 45, hjust = 1)
    )
```

```{r fig2c.save, fig.width = 2.5, fig.height = 3}
fig2c
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "2C.PsychENCODE_development", .x), width = fig_width, height = 3)
)
```

## D | GRCCA GO enrichments

```{r Fig2d.prep}
## Determine labels
fig2d_labels <- c(
    "cilium movement",
    "motile cilium",
    "axoneme",
    #"axonemal microtubule",
    "axoneme assembly",
    "adaptive immune response",
    "outer dynein arm assembly",
    "MHC class II protein complex",
    "T cell costimulation",
    "MHC class II protein complex binding",
    "minus-end-directed microtubule motor activity",
    "dynein light intermediate chain binding",

    "syntaxin-1 binding",
    "ubiquitin-dependent protein catabolic process",
    "endosome to lysosome transport",
    "cytoplasmic translation",
    "regulation of autophagy",
    "postsynaptic density",
    "presynaptic endocytic zone membrane",
    "ATP binding",
    "ATP hydrolysis activity",
    "protein ubiquitination"
)

## Filter for terms of interest & assign labels
df_fig2d <- df_grcca_fgsea_go %>% 
    dplyr::filter(padj < 0.05) %>% 
    mutate(direction = ifelse(NES < 0, "down", "up")) %>% 
    group_by(direction) %>% 
    mutate(label = ifelse(term %in% fig2d_labels, str_wrap(term, 25), ""))
```

```{r Fig2d.plot}
fig2d <- df_fig2d %>%
    ggplot(aes(x = padj, y = NES)) +
    geom_point(aes(size = size, fill = NES), color = "gray", shape = 21) +
    geom_text_repel(aes(label = label), size = 2.75, force = 20,
                    box.padding = 0.25, min.segment.length = 0, max.overlaps = 30) +
    geom_hline(yintercept = 0, color = "gray") +
    facet_wrap(vars(ontology)) +
    scale_fill_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-3, 3), guide = "none") +
    scale_size_continuous(range = c(2, 6)) +
    scale_x_reverse(breaks = c(0, 0.03, 0.05)) +
    coord_cartesian(clip = "off") +
    guides(size = guide_legend(title = "n genes in path", title.hjust = 0.5, title.position = "top")) +
    labs(x = "Enrichment FDR", y = "NES",
         title = "D | Functional pathway enrichments") +
    theme(#legend.position = "bottom",
          #legend.justification = "center",
          title.justification = "center",
          legend.key.size = unit(0.3, "cm")
    )
```

```{r Fig2d.save, fig.height = 3, fig.width = 6}
fig2d
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "2D.GO_enrichment", .x), width = 6, height = 3.0)
)
```

# Fig 3

## A | Top n genes by L2FC t-statistic

```{r Fig3a.plot}
fig3a <- df_de_res %>% 
    slice_max(order_by = abs(stat), n = 10) %>% 
    
    ggplot(aes(x = stat, y = reorder(gene_symbol, stat))) +
    geom_point(aes(size = -log10(pvalue), fill = stat), shape = 21) +
    geom_vline(xintercept = 0, color = "gray") +
    scale_fill_gradientn(colors = gene_weight_color_scale, 
                         limits = c(-5, 5), guide = "none") +
    scale_size_continuous(range = c(1, 4)) +
    guides(size = guide_legend(title = "-log10(p-value)")) +
    xlim(c(-5, 5)) +
    coord_cartesian(clip = "off") +
    labs(y = NULL, x = "t-statistic", 
         title ="A | Top 10 DEGs"
    ) +
    theme(legend.position = c(0.05, 0.7), 
          legend.box = "horizontal",
          legend.key.size = unit(0.35, "cm")
    )
```

```{r Fig3a.save, fig.width = 2.5, fig.height = 2}
fig3a
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "3A.top_10_DEGs", .x), width = fig_width, height = fig_height)
)
```

## B | GRCCA structure correlations (rx) vs DGE effect size

```{r Fig3b.plot}
fig3b <- df_x_res %>% 
    left_join(df_de_res) %>% 
    
    ggplot(aes(x = stat, y = pearsons_r)) +
    geom_hline(yintercept = 0, color = "gray") +
    geom_vline(xintercept = 0, color = "gray") +
    geom_point(aes(color = pearsons_r + stat), size = 0.5) +
    geom_smooth(method = "lm", color = "black", linewidth = 0.5) +
    stat_cor(aes(label = after_stat(r.label)), label.sep = "\n", size = 3,
             label.y.npc = "top", vjust = 0) +
    scale_color_gradientn(colors = gene_weight_color_scale, limits = c(-5.4, 5.4)) +
    coord_cartesian(clip = "off") +
    #scale_x_continuous(breaks = c(-0.2, 0, 0.2)) +
    labs(x = "DGE t-statistic", y = expression("Structure correlation (" * italic(r)[x] * ")"),
         title = "B | DGE vs GRCCA results"
    ) +
    theme(legend.position = "none")
```

```{r Fig3b.save, fig.width = 2.5, fig.height = 2}
fig3b
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "3B.gene_strucCor_vs_DE_stat", .x), width = fig_width, height = fig_height)
)
```

## C | Compare DGE and GRCCA functional enrichments

```{r Fig3c.prep}
df_fig3c <- df_grcca_fgsea_go %>% 
    mutate(analysis = "GRCCA", .before = 1) %>% 
    bind_rows(df_de_fgsea_go %>% 
                  mutate(analysis = "DGE", .before = 1)
    ) %>% 
    mutate(class = ifelse(NES < 0, "NES < 0", "NES > 0")) %>% 
    filter(padj < 0.05) %>% 
    count(analysis, class) %>% 
    mutate(text_color = ifelse(analysis == "DGE", "white", "black"),
           analysis_color = ifelse(analysis == "GRCCA", "white", "black")
    )
```

```{r Fig3c.plot}
fig3c <- df_fig3c %>% 
    ggplot(aes(x = analysis, y = n)) +
    geom_rect(aes(xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf, fill = class), alpha = 0.35) +
    geom_col(aes(fill = I(analysis_color)), color = "black") +
    geom_text(aes(label = n, color = I(text_color)), size = 3, vjust = 1.2) +
    facet_wrap(vars(class)) +
    scale_fill_manual(values = c("NES < 0" = gene_weight_color_scale[15], "NES > 0" = gene_weight_color_scale[85])) +
    labs(x = NULL, y = "n pathways (FDR < 0.05)",
         title = "C | DGE vs GRCCA GO results"
    ) +
    theme(
        #panel.background = element_rect(fill = "transparent"),
        legend.position = "none",
        axis.text = element_text(size = 8)
    )
```

```{r Fig3c.save, fig.width = 2.5, fig.height = 2}
fig3c
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "3C.GRCCA_vs_DGE_GO", .x), width = 2.5, height = 2)
)
```

## D | DEG enrichment for GWAS results

```{r Fig3d.prep}
df_fig3d <- df_de_fgsea_benchmark %>% 
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), "SCZ", "other") %>% factor(levels = c("SCZ", "other"))) %>% 
    mutate(benchmark_list = str_remove_all(benchmark_list, "SCZ|\\)|\\(") %>% 
               str_trim %>% 
               str_replace(", ", ",\n")
    ) %>% 
    mutate(label_color = ifelse(padj < 0.05, "white", "black")) %>% 
    mutate(box_color = ifelse(padj < 0.05, "black", "transparent")) %>% 
    mutate(label = round(NES, 2) %>% format(digits = 3))
```

```{r Fig3d.plot}
fig3d <- df_fig3d %>% 
    ggplot(aes(x = benchmark_list, y = 1)) +
    geom_tile(aes(fill = NES, color = I(box_color), alpha = abs(NES)), width = 0.95, height = 0.95, linewidth = 0.75) +
    geom_text(aes(label = label, color = I(label_color)), size = 2.75) +
    facet_wrap(vars(class), scales = "free", nrow = 2) +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1.58, 1.58)) +
    scale_alpha_continuous(guide = "none", limits = c(0.83, 1.58)) +
    guides(fill = guide_colorbar(title.hjust = 0.5, title.position = "left")) +
    labs(x = NULL, y = NULL,
         title = "D | Risk gene enrichment") +
    theme(axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_blank(),
          axis.text.x = element_text(margin = margin(t = -3)),
          strip.background = element_rect(fill = "white", color = "black"),
          strip.text = element_text(size = 9),
          legend.margin = margin(r = -15),
          legend.position = "left",
          legend.justification = "top",
          legend.title = element_text(angle = 90),
          legend.key.height = unit(0.60, "cm"),
          legend.key.width = unit(0.15, "cm")
    )
```

```{r Fig3d.save, fig.width = 2.5, fig.height = 2.0}
fig3d
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(poster_figures_dir, "3D.Risk_gene_enrichment", .x), width = fig_width, height = fig_height)
)
```


