---
title: "Supplemental figures for RLSmith Transl Psych 2025"
author: "Rachel L Smith"
output: pdf_document
---

#----------------FOLLOWING FIRST ROUND REVISIONS-------------------------------#

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width = 2.5, fig.height = 2, collapse = TRUE, warning = FALSE, message = FALSE, echo = FALSE)

## Set paths and load data
base_dir <- "~/Documents/PhD/projects/sgacc_wgcna_grcca/"
supplement_figures_dir <- paste0(base_dir, "outputs/figures/supplement/panels/")
tables_dir <- paste0(base_dir, "outputs/tables/")
analysis_objects_dir <- paste0(base_dir, "outputs/objects/")

source(paste0(base_dir, "scripts/load_WGCNA_res.R"))

## Load data objects for plots
objects <- list.files(analysis_objects_dir)
objects <- objects[!str_detect(objects, "null|archive")]
for (obj in objects){
  print(paste0("loading ", obj, "...."))
  load(paste0(analysis_objects_dir, obj))
}
```

# TABLES #

```{r TableS1}
write_xlsx(df_covariates %>% 
               mutate(dx = factor(dx, levels = names(dx_colors))) %>% 
               arrange(dx, sex_at_birth, race, age_death) %>% 
               mutate_at(vars(c(bmi, height, weight)), ~ as.numeric(.x)) %>% 
               dplyr::select(-suicide, -sample), 
           path = paste0(tables_dir, "TableS1.covariate_information.xlsx"))
```

```{r TableS2}
write_xlsx(list("A.GRCCA results" = df_results %>% dplyr::select(-set) %>% dplyr::select(varx, everything()),
                "B.Covariate (Y) results" = df_y_res %>% dplyr::rename("structure_correlation" = "pearsons_r"),
                "C.Gene (X) results" = df_x_res %>% dplyr::rename("structure_correlation" = "pearsons_r")
                ), 
           path = paste0(tables_dir, "TableS2.GRCCA_results.xlsx")
)
```

```{r TableS3}
write_xlsx(
    list(
        "A.Risk genes" = df_grcca_fgsea_benchmark %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "B.Cell-type" = df_grcca_fgsea_cell %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "C.GO" = df_grcca_fgsea_go %>% 
            dplyr::filter(padj < 0.05) %>%
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes))
    ), 
    path = paste0(tables_dir, "TableS3.GRCCA_GSEA_enrichments.xlsx")
)
```

```{r TableS4}
write_xlsx(
    list("TableS4" = df_limma_res %>% arrange(-abs(t))), 
    path = paste0(tables_dir, "TableS4.DEGs.xlsx")
)
```

```{r TableS5}
### Export enrichment results as table ###
write_xlsx(
    list(
        "A.Cell-type" = df_de_fgsea_cell %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "B.GO" = df_de_fgsea_go %>% 
            dplyr::filter(padj < 0.05) %>%
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "C.Risk genes" = df_de_fgsea_benchmark %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes))
    ), 
    path = paste0(tables_dir, "TableS5.DEG_GSEA_enrichments.xlsx")
)
```

```{r TableS6}
write_xlsx(list("A.GRCCA results" = df_results_transcripts %>% dplyr::select(-set) %>% dplyr::select(varx, everything()),
                "B.Covariate (Y) results" = df_y_res_transcripts %>% dplyr::rename("structure_correlation" = "pearsons_r"),
                "C. Transcript (X) results" = df_x_res_transcripts %>% dplyr::rename("structure_correlation" = "pearsons_r")), 
           path = paste0(tables_dir, "TableS6.GRCCA_transcript_results.xlsx")
)
```

```{r TableS7}
write_xlsx(list(
    "A. Module gene assignments" = df_modules_filt %>% left_join(df_ensembl_to_symbol) %>% dplyr::select(module, color, ensembl_gene_id, gene_symbol) %>% as.data.frame,
    "B. Module GO results" = df_mods_go %>% dplyr::filter(pvalue < 0.05 & module != 0) %>% dplyr::select(-qvalue) %>% as.data.frame
),
path = paste0(tables_dir, "TableS7.WGCNA_module_results.xlsx")
)
```

```{r TableS8}
write_xlsx(df_kme %>% 
               dplyr::filter(p_value < 0.05 & pc == "PC1") %>% 
               dplyr::select(-c(data, pc, proportion_of_variance, estimate, std_error)),
           path = paste0(tables_dir, "TableS8.kME_analysis_significant_results.xlsx")
)
```

```{r TableS9}
write_xlsx(list("A.RCCA results" = df_results_rcca %>% dplyr::select(-set) %>% dplyr::select(varx, everything()),
                "B.Covariate (Y) results" = df_y_res_rcca %>% dplyr::rename("structure_correlation" = "pearsons_r"),
                "C.Gene (X) results" = df_x_res_rcca %>% dplyr::rename("structure_correlation" = "pearsons_r")), 
           path = paste0(tables_dir, "TableS9.RCCA_results.xlsx")
)
```

```{r TableS10}
write_xlsx(
    list(
        "A.Cell-type" = df_rcca_fgsea_cell %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "B.GO" = df_rcca_fgsea_go %>% 
            dplyr::filter(padj < 0.05) %>%
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes)),
        "C.Risk genes" = df_rcca_fgsea_benchmark %>% 
            mutate(genes = map_chr(leadingEdge, ~ f_collapse_genes(.x, index = 30000))) %>% 
            dplyr::select(-leadingEdge) %>% unnest(cols = c(genes))
    ), 
    path = paste0(tables_dir, "TableS10.RCCA_GSEA_enrichments.xlsx")
)
```


# FIG S1: Transcript filtering by coefficient of variation #

### A | Relationship between mean and standard deviation
```{r FigS1a.plot, fig.width = 2.6}
df_mean_sd %>% 
  ggplot(aes(x = mean %>% log10, y = stdev %>% log10)) +
  geom_point(shape = 21, size = 0.75) +
  #geom_density2d(linewidth = 0.25) +
  geom_smooth(method = "lm", color = "maroon", linewidth = 0.5) +
  stat_cor(label.sep = "\n", vjust = 1, size = 3.5) +
  labs(x = "log10(mean)", y = "log10(standard deviation)",
       title = "A | Mean vs standard deviation \nin transcript raw counts")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S1A.transcript_raw_counts_mean_stdev", .x), width = 2.6, height = fig_height)
)
```

### B | Relationship between mean and coefficient of variation
```{r FigS1b.plot, fig.width = 2.6}
df_mean_sd %>% 
    ggplot(aes(x = log10(mean), y = CV %>% log10)) +
    geom_point(shape = 21, size = 0.75) +
    geom_density2d() +
    geom_smooth(method = "lm", color = "maroon", linewidth = 0.5) +
    stat_cor(label.sep = "\n", label.x.npc = "right",
             vjust = 1, hjust = 1, size = 3.5
    ) +
    labs(x = "log10(mean)", y = "log10(coefficient of variation)",
         title = "B | Mean vs coefficient of \nvariation in transcript counts")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S1B.transcript_raw_counts_mean_CV", .x), width = 2.6, height = fig_height)
)
```

### C | Cutoff for coefficient of variation
```{r FigS1c.plot}
df_mean_sd %>% 
  ggplot(aes(x = CV)) +
  geom_hline(yintercept = 0) +
  geom_point(aes(y = 3.5), position = position_jitter(0.001), alpha = 0.5, size = 0.75, shape = 21) +
  geom_density(color = "blue", fill = "transparent") +
  geom_boxplot(aes(y = -0.1), color = "blue", width = 0.2, outlier.shape = NA) +
  geom_vline(xintercept = cutoff, lty = 2, color = "maroon") +
  labs(x = "Coefficient of variation (CV)", y = NULL,
       title = "C | Coefficient of variation \nacross transcripts") +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank()
  ) 
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S1C.CV_distribution", .x), width = fig_width, height = fig_height)
)
```


# FIG S2: Raw count preprocessing: qSV and known covariate regression #

### A | Variance explained by qSVs
```{r FigS2a.plot}
df_qsv_var_expl %>% #dplyr::count(feature)
    dplyr::filter(feature == "Proportion of Variance" & qSV %in% 1:16) %>%
    
    ggplot(aes(x = qSV, y = value*100)) +
    geom_col(aes(fill = value), color = "black") +
    geom_hline(yintercept = 0.02*100, color = "maroon") +
    scale_fill_gradient(low = "white", high = "black", 
                        limits = c(0, 0.06), guide = "none") +
    #scale_y_continuous(breaks = seq(0, 0.06, 0.01)) +
    coord_cartesian(clip = "off") +
    labs(y = "% variance explained",
         title = "A | qSV variance explained") +
    theme(axis.text.x = element_text(angle = 45))
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S2A.qSV_variance_explained", .x), width = fig_width, height = fig_height)
)
```

### B | Correlate qSVs with known covariates
```{r FigS2b.prep}
## Define covariate order based on clustering within each type
f_define_order <- function(covariate_matrix, covariate_type = c("technical", "biological", "toxicological")) {
    if (covariate_type == "technical") {
        subset <- technical_covariates
    } else if (covariate_type == "biological") {
        subset <- biological_covariates
    } else if (covariate_type == "toxicological") {
        subset <- drug_covariates
    } else (stop("Covariate type must be technical, biological, or toxicological"))
    matrix_tmp <- covariate_matrix[,subset]
   order <- matrix_tmp[hclust(dist(matrix_tmp))$order,] %>% colnames
   return(order)
}
m_covariates <- df_covariates_numeric %>% dplyr::select(technical_covariates, biological_covariates, drug_covariates) %>%  as.matrix()
covariate_order <- c(
    f_define_order(m_covariates, "biological"),
    f_define_order(m_covariates, "toxicological"),
    f_define_order(m_covariates, "technical")
)
covariate_order


## Switch major_stimulants_cocaine_included and cannabinoids for the plot
swtch <- function(vec, value1, value2) {
  # Ensure the values exist in the vector
  if (!(value1 %in% vec) || !(value2 %in% vec)) {
    stop("Both values must exist in the vector.")
  }
  
  # Find the positions of the values
  pos1 <- which(vec == value1)
  pos2 <- which(vec == value2)
  
  # Swap the elements
  vec[pos1] <- value2
  vec[pos2] <- value1
  
  return(vec)
}
covariate_order <- swtch(covariate_order, "major_stimulants_cocaine_included", "cannabinoids")

## Create dataframe defining covariate type & order
df_figS2b <- df_covariate_qsv_cor %>% 
    mutate(significance = case_when(
        p_adj < 0.05 & p_adj > 0.01 ~ "*",
        p_adj < 0.01 ~ "**",
        TRUE ~ ""
    ),
           #covariate = factor(covariate, levels = covariate_order),
           type = case_when(
               covariate %in% technical_covariates ~ "Technical",
               covariate %in% biological_covariates ~ "Biological",
               covariate %in% drug_covariates ~ "Toxicological",
               TRUE ~ "Diagnostic group"
           ) %>% 
               factor(levels = c("Diagnostic group", "Technical", "Toxicological", "Biological"))
    ) %>% 
    filter(type != "Diagnostic group") %>% 
    arrange(type, qsv, pearsons_r) %>% 
    mutate(covariate = factor(covariate, levels = covariate_order))
```

```{r FigS2b.plot, fig.height = 6.5, fig.width = 4}
df_figS2b %>% 
    ggplot(aes(x = qsv, y = covariate)) +
    geom_tile(aes(fill = pearsons_r), width = 0.98, height = 0.98) +
    geom_text(aes(label = significance), vjust = 0.75) +
    facet_wrap(vars(type), scales = "free", ncol = 1) +
    force_panelsizes(rows = c(length(biological_covariates), length(drug_covariates), length(technical_covariates))) +
    
    scale_fill_gradientn(colors = rev(brewer.rdbu(100))) +
    coord_cartesian(clip = "off") +
    guides(fill = guide_colorbar(title = "Pearson's r", title.hjust = 0.5, title.position = "left")) +
    labs(x = "qSV", y = NULL, 
         title = "B | qSVs and known \ncovariate correlations"
    ) +
    theme(legend.position = c(-1.1, 1.0),
          legend.key.height = unit(0.6, "cm"),
          legend.key.width = unit(0.15, "cm"),
          legend.title = element_text(angle = 90),
          axis.text.x = element_text(angle = 45),
          strip.background = element_rect(fill = "white", color = "black")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S2B.qSV_covariate_correlations", .x), width = 4, height = 6.5)
)
```

### C | PC-covariate correlations before and after covariate regressions

```{r FigS2c.prep}
df_figS2c <- df_pca1_covar_cor %>% 
    mutate(regression = "Before", .before = 1) %>% 
    bind_rows(
        df_pca3_covar_cor %>% 
            mutate(regression = "After", .before = 1) 
    ) %>% 
    mutate(regression = factor(regression, levels = c("Before", "After"))) %>% 
    filter(covariate != "dx") %>% 
    mutate(covariate = factor(covariate, levels = c(covariate_order, paste0("qSV", 1:16))),
           significance = case_when(
               p_adj < 0.05 & p_adj > 0.01 ~ "*",
               p_adj < 0.01 ~ "**",
               TRUE ~ ""
           ),
           type = case_when(
               covariate %in% technical_covariates ~ "Technical",
               covariate %in% biological_covariates ~ "Biological",
               covariate %in% drug_covariates ~ "Toxicological",
               TRUE ~ "qSV"
           ) %>% 
               factor(levels = c("qSV", "Biological", "Toxicological", "Technical"))
    )
```

```{r FigS2c.plot, fig.width = 6.5, fig.height = 4}
df_figS2c %>% 
    ggplot(aes(x = pc, y = covariate)) +
    geom_tile(aes(fill = pearsons_r), width = 0.98, height = 0.98, linewidth = 0.5) +
    geom_text(aes(label = significance), vjust = 0.75, angle = 90) +
    
    facet_grid2(regression ~ type, scales = "free") +
    force_panelsizes(rows = c(1, 4), cols = c(16, length(biological_covariates), length(drug_covariates), length(technical_covariates))) +
    scale_fill_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-0.43, 0.43)) +
    scale_color_manual(values = c("yes" = "black", "no" = "transparent"), guide = "none") +
    coord_flip(clip = "off") +
    guides(fill = guide_colorbar(title = "Pearson's r", title.hjust = 0.5, title.position = "top")) +
    labs(x = "PC (> 2% explained var)", y = NULL,
         title = "C | PC-covariate correlations before and after regression") +
    theme(legend.position = c(0, -0.4),
          legend.direction = "horizontal",
          legend.key.width = unit(0.6, "cm"),
          legend.key.height = unit(0.15, "cm"),
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 8),
          strip.background = element_rect(fill = "white", color = "black"),
          strip.clip = "off",
          panel.spacing = unit(0.1, "lines")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S2C.PCA_covariate_cor", .x), width = 6.5, height = 4)
)
```

# FIG S3: Toxicology Multiple Correspondence Analysis results #

### A | Matrix of drug use across 185 samples

```{r FigS3a.prep}
## Format drug names for publications
df_drugs_numeric <- df_covariates_numeric %>% 
  dplyr::select(sample, dx, all_of(drug_covariates)) #%>% 
  
  # # change drug names for figures
  # dplyr::rename_all(~str_replace_all(.x, "anti_", "anti")) %>% 
  # dplyr::rename_all(~str_replace_all(.x, "_", " ")) %>% 
  # dplyr::rename(
  #   "nicotine/cotinine" =  "nicotine cotinine",
  #   "sedatives, hypnotics, anxiolitics" =  "sedative hypnotic anxiolitics",
  #   "major stimulants (cocaine included)" =  "major stimulants cocaine included",
  #   "benzodiazepines" = "benzos",
  #   "other psychotropic drug" = "other psychotropic drug"
  # )

## Cluster samples to arrange for plot
samples_order <- df_drugs_numeric %>% 
  group_by(dx) %>% 
  nest() %>% 
  mutate(order = map(
    .x = data,
    .f = ~ .x %>% 
      column_to_rownames("sample") %>% 
      dist %>% 
      hclust %>% 
      .$order
  )) %>% 
  unnest(cols = c(data, order)) %>% 
  dplyr::select(dx, order, sample) %>% 
  mutate(dx = case_when(
    str_detect(sample, "control") ~ 1,
    str_detect(sample, "bipolar") ~ 2,
    str_detect(sample, "mdd") ~ 3,
    str_detect(sample, "schizo") ~ 4
  )) %>% 
  arrange(dx, order) %>% 
  pull(sample)

## Cluster drugs by sample use
drugs_order <- f_define_order(m_covariates, "toxicological")

## Format tibble with arrange rows & columns
df_figS3a <- df_drugs_numeric %>% 
    pivot_longer(3:ncol(.), names_to = "drug", values_to = "value") %>% 
    mutate(value = case_when(
        value == 1 ~ "detected",
        value == 0 ~ "not detected",
        is.na(value) ~ "unknown"
    ), 
    sample = factor(sample, levels = samples_order),
    drug = factor(drug, levels = drugs_order),
    dx = case_when(
        dx == 0 ~ "BD",
        dx == 1 ~ "Control",
        dx == 2 ~ "MDD",
        dx == 3 ~ "SCZ"
    ) %>% factor(levels = names(dx_colors))
    )

## Set matrix colors
matrix_colors <- c(
    "detected" = "white",
    "not detected" = "black",
    "unknown" = "gray"
)
```

```{r FigS3a.plot, fig.height = 2.25, fig.width = 4.5}
df_figS3a %>%
  ggplot(aes(x = sample)) +
  geom_tile(aes(y = drug, fill = value)) +
  geom_tile(aes(y = 0, fill = dx)) +
  geom_vline(xintercept = c(55.5, 90.5, 141.5), color = "maroon", linewidth = 0.5) +
  scale_fill_manual(values = c(matrix_colors, dx_colors),
                    na.value = "gray") +
  guides(fill = guide_legend(title = NULL)) +
  labs(y = NULL, title = "A | Sample toxicology data") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = c(-0.85, 0.92),
        legend.key.size = unit(0.2, "cm"),
        legend.text = element_text(size = 8),
        axis.line = element_blank()
        )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S3A.sample_toxicology_matrix", .x), width = 4.5, height = 2.25)
)
```

### B | Variance explained of each drug MCA dimension

```{r FigS3b.plot, fig.width = 2.1}
df_eig %>% 
    mutate(dim = factor(dim),
           label = paste0(round(cum_var, 2), "%")) %>% 
    
    ggplot(aes(x = dim)) +
    geom_col(aes(y = var, fill = var), color = "black") +
    #geom_point(aes(y = cum_var)) +
    # geom_text(aes(y = cum_var, label = label),
    #           size = 3, vjust = -1) +
    geom_hline(yintercept = 5, color = "maroon") +
    scale_fill_gradient(low = "white", high = "black", guide = "none") +
    labs(title = "B | MCA dimension \nvariance explained",
         x = "Dimension", y = "% variance explained") +
    theme(axis.text.x = element_text(angle = 45, size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S3B.MCA_variance_explained", .x), width = 2.1, height = 2)
)
```

### C | MCA variable loadings

```{r FigS3C.prep}
df_figS3c <- df_var_loadings %>% 
  pivot_longer(3:ncol(.), names_to = "dimension", values_to = "score") %>% 
  filter(use == "yes") %>% 
  mutate(drug = factor(drug, levels = drugs_order),
         dimension = str_replace(dimension, "dim", "") %>% factor(levels = 1:17),
         label = ifelse(abs(score) > 0.015, "*", NA)
  )
```

```{r FigS3C.plot, fig.width = 3.75, fig.height = 3}
df_figS3c %>% 
    ggplot(aes(x = drug, y = dimension)) +
    geom_tile(aes(fill = score)) +
    geom_text(aes(label = label), vjust = 0.75) +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-0.065, 0.065)) +
    guides(fill = guide_colorbar(title = "Loading", title.position = "left", title.hjust = 0.5)) +
    coord_flip() +
    labs(y = "Dimension", x = NULL, 
         title = "C | Variable loadings on \n MCA dimensions") +
    theme(legend.position = c(-1.15, 0.92),
          legend.key.width = unit(0.15, "cm"),
          legend.key.height = unit(0.60, "cm"),
          legend.title = element_text(angle = 90),
          axis.text.y = element_text(size = 8.5),
          axis.text.x = element_text(angle = 45, hjust = 1, size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S3C.MCA_variable_loadings", .x), width = 3.75, height = 3)
)
```


### D | MCA dimension - covariate correlations

```{r FigS3d.prep}
## Arrange covariates and format for plotting
df_figS3d <- df_mca_covar_cor %>% 
    # remove technical covariates & dx
    filter(!(covariate %in% technical_covariates) & covariate != "dx") %>% 
    mutate(
        covariate = factor(covariate, levels = covariate_order),
        dimension = str_remove(dimension, "dim") %>% as.numeric %>% as.factor,
        #significant = ifelse(p_adj < 0.05, "yes", "no")
        significance = case_when(
            p_adj < 0.05 & p_adj > 0.01 ~ "*",
            p_adj < 0.01 ~ "**",
            TRUE ~ ""
        ),
        type = case_when(
               #covariate %in% technical_covariates ~ "Technical",
               covariate %in% biological_covariates ~ "Biological",
               covariate %in% drug_covariates ~ "Toxicological",
               TRUE ~ "Diagnostic group"
           ) %>% 
               factor(levels = c("Diagnostic group", "Biological", "Toxicological"))
    )
```

```{r FigS3d.plot, fig.width = 3.75, fig.height = 5}
df_figS3d %>% filter(type != "Diagnostic group") %>% 
    
    ggplot(aes(x = dimension, y = covariate)) +
    geom_tile(aes(fill = pearsons_r), linewidth = 0.5) +
    geom_text(aes(label = significance), vjust = 0.75, size = 2.5) +
    facet_wrap(vars(type), scales = "free_y", ncol = 1) +
    force_panelsizes(rows = c(length(biological_covariates), length(drug_covariates))) +
    scale_fill_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-1.0, 1.0)) +
    scale_color_manual(values = c("yes" = "black", "no" = "transparent"), guide = "none") +
    guides(fill = guide_colorbar(title = "Pearson's r", title.position = "left", title.hjust = 0.5)) +
    labs(x = "Dimension", y = NULL, 
         title = "D | MCA dimension \ncorrelations with \nknown covariates") +
    theme(
        legend.position = c(-0.80, 1),
        legend.key.height = unit(0.6, "cm"),
        legend.key.width = unit(0.15, "cm"),
        legend.title = element_text(angle = 90),
        axis.text.y = element_text(size = 8.5),
        axis.text.x = element_text(angle = 45, hjust = 1)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S3D.MCA_dimension_covariate_correlations", .x), width = 3.75, height = 5)
)
```


# FIG S4: WGCNA module sizes across parameter space #

### A | Soft-threshold power selection

```{r FigS4a.plot, fig.width = 5, fig.height = 2}
## Model fit of each soft thresholding power
p1 <- df_sft %>% 
    ggplot(aes(x = power, y = -sign(slope)*sft_r_sq)) +
    geom_point(aes(color = power == 3), size = 1) +
    #geom_text(aes(label = power), size = 3) +
    geom_hline(yintercept = 0.9, lty = 2, color = "maroon") +
    scale_color_manual(values = c("TRUE" = "maroon", "FALSE" = "black"), guide = "none") +
    scale_x_continuous(breaks = seq(2, 16, by = 2)) +
    labs(
        x = "Soft threshold (power)",
        y = "Scale free topology model fit \n(signed R^2)"
    ) +
    theme(
        panel.grid.major.x = element_line(color = "darkgrey", size = 0.2)#,
        #panel.grid.minor.x = element_line(color = "lightgray")
    )

## Median connectivity as a function of the soft-thresholding power (want to remain reasonably high (in the 100s) or above)
p2 <- df_sft %>% 
    ggplot(aes(x = power, y = median_k)) +
    geom_point(aes(color = power == 3), size = 1) +
    #geom_text(aes(label = power), size = 3) +
    geom_hline(yintercept = 100, lty = 2, color = "maroon") +
    scale_color_manual(values = c("TRUE" = "maroon", "FALSE" = "black"), guide = "none") +
    scale_x_continuous(breaks = seq(2, 16, by = 2)) +
    labs(
        x = "Soft threshold (power)",
        y = "Median connectivity"
    ) +
    theme(
        panel.grid.major.x = element_line(color = "darkgrey", size = 0.2)#,
        #panel.grid.minor.x = element_line(color = "lightgray")
    )

## Combine and save
p1 + p2 + plot_annotation(title = "A | Soft-thresholding power fit to data")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S4A.SFT_fit", .x), width = 5, height = fig_height)
)
```

### B | Module stability across minimum module size & cut height thresholds

```{r FigS4b.plot, fig.width = 5, fig.height = 6.5}
df_modules %>%
    group_by(sft, min_size, cut_height) %>%
    dplyr::count(module, color) %>% 
    #filter(cut_height != 1) %>%
    
    ggplot(aes(x = module)) +
    geom_segment(aes(y = 0, yend = log10(n), color = color)) +
    geom_point(aes(y = log10(n), fill = color), shape = 21, size = 0.75) +
    geom_vline(aes(xintercept = 21), lty = 2, color = "maroon") +
    scale_fill_identity() +
    scale_color_identity() +
    scale_x_discrete(breaks = seq(0, 50, 10)) +
    facet_grid(cut_height ~ min_size) +
    labs(y = "log10(N genes)", 
         title = "B | Module assignments across WGCNA parameters"
    ) +
    theme(
        strip.text = element_text(size = 9),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S4B.module_sizes", .x), width = 5, height = 6.5)
)
```


# FIG S5: WGCNA module enrichments #

### A | Module sizes

```{r FigS5a.plot, fig.width = 3.5, fig.height = 2.5}
df_modules_filt %>% 
    dplyr::count(module, color) %>% 
    ggplot(aes(x = module, y = n)) +
    geom_col(aes(fill = module), color = "black", linewidth = 0.25) +
    geom_text(aes(label = n), size = 3, angle = 45, hjust = 0, vjust = 0) +
    scale_fill_manual(values = module_colors) +
    coord_cartesian(clip = "off") +
    labs(title = "A | Module sizes",
         x = NULL, y = "Number of genes") +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 0.9, size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S5A.Module_sizes", .x), width = 3.5, height = 2.5)
)

```

### B | Semantic themes across module GO results

```{r FigS5b.prep}
n_topics <- topic_model$settings$dim$K
## Extract topic descriptions (frequent words)
# FREX weights words by frequency and exclusivity to the topic
df_word_beta <- tidy(topic_model, matrix = "beta") %>%
    group_by(topic) %>%
    arrange(topic, -beta) %>% 
    top_n(n = 10, wt = beta) %>%
    mutate(topic = paste0("topic ", topic)) %>% 
    dplyr::filter(!str_detect(term, "cell")) %>% 
    group_by(topic) %>% 
    mutate(top_word = ifelse(beta == max(beta), "yes", "no"))

## Extract topic probabilities for each module (document)
df_group_gamma <- tidy(
    topic_model, 
    matrix = "gamma",
    document_names = rownames(go_sparse)
) %>% 
    mutate(module = factor(document)) %>% 
    dplyr::select(module, topic, gamma) %>% 
    mutate(topic = factor(topic, levels = seq(n_topics, 1, by = -1)),
           module = paste0("geneM", module) %>% factor(levels = paste0("geneM", levels(module)))
    ) %>% 
    dplyr::filter(gamma > 0.01)
```

```{r FigS3b.plot, fig.width = 3, fig.height = 6}
## Plot word strength of association with each topic
p_figS5b_beta <- df_word_beta %>% 
    ggplot(aes(x = reorder_within(term, beta, topic), y = beta, fill = beta)) +
    geom_col(color = "#222222") +
    geom_text(aes(label = term, y = 0.001, size = beta, color = top_word), angle = 90, hjust = 0) +
    facet_wrap(vars(topic), ncol = 1, scales = "free_x") +
    scale_x_reordered() +
    scale_y_continuous(limits = c(0, 0.06), breaks = c(0, 0.05)) +
    scale_size_continuous(range = c(2, 3), guide = "none") +
    scale_color_manual(values = c("yes" = "white", "no" = "black"), guide = "none") +
    scale_fill_gradient(low = "white", high = "blue", guide = "none", limits = c(0, 0.058)) +
    labs(x = NULL, y = "Word strength of association with each topic") +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_rect(fill = "white", color = "black"),
          strip.text = element_text(size = 9)
    )

## Plot module strength of association with each topic
p_figS5b_gamma <- df_group_gamma %>%
    ggplot(aes(x = gamma, y = topic, color = module)) +
    geom_jitter(aes(alpha = gamma), position = position_jitter(seed = 2, width = 0.15)) +
    geom_text_repel(aes(alpha = gamma, label = module), size = 2.75, box.padding = 0.2, min.segment.length = 0,
                    max.overlaps = 15, position = position_jitter(seed = 2, width = 0.15)) +
    geom_hline(yintercept = c(1.5, 2.5, 3.5, 4.5, 5.5), color = "black", linewidth = 0.25) +
    scale_color_manual(values = module_colors) +
    labs(x = "Module strength of association \nwith each topic", y = NULL) +
    theme(legend.position = "none",
          axis.text.y = element_blank()
    )

## Combine & save
p_figS5b_beta + p_figS5b_gamma + plot_layout(widths = c(1, 2)) + plot_annotation(title = "B | GO semantic themes")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S5B.topic_modeling", .x), width = 3, height = 6)
)
```


### C | Cell-type enrichment of GO modules

```{r FigS5c.prep}
df_figS5c <- df_mods_cell_type_hypergeometric %>%
    mutate(
        p_adj = ifelse(p_adj == 0, 10^-15, p_adj),
        color = ifelse(p_value < 0.05, paste0(module), NA) %>% factor,
        label = case_when(
            p_adj < 0.05 & p_adj > 0.01 ~ "*",
            p_adj < 0.01 & p_adj > 0.001 ~ "**",
            p_adj < 0.001 ~ "***",
            TRUE ~ ""
        )
    )
```

```{r FigS5c.plot, fig.width = 3.35, fig.height = 3.4}
df_figS5c %>% 
    ggplot(aes(x = cell_type, y = module)) +
    geom_tile(aes(fill = color, alpha = -log10(p_adj), color = color), 
              width = 0.95, height = 0.95, linewidth = 0.5) +
    geom_text(aes(label = label), vjust = 0.75, size = 3.5) +
    #annotate(geom = "text", label = "* = FDR < 0.05", x = 6.5, y = 1, size = 3) +
    scale_alpha_continuous(range = c(0.1, 1), na.value = 1) +
    scale_fill_manual(values = module_colors, na.value = "transparent", guide = "none") +
    scale_color_manual(values = module_colors, na.value = "transparent", guide = "none") +
    guides(alpha = guide_legend(title = "-log10(FDR)", title.position = "top", title.hjust = 0.5)) +
    labs(title =  "C | Cell-type enrichment", 
         x = NULL, y = NULL
    ) +
    theme(legend.position = "bottom",
          legend.margin = margin(t = -12),
          legend.key.size = unit(0.3, "cm"),
          legend.justification = "center",
          axis.text.y = element_text(size = 8.5),
          axis.text.x = element_text(angle = 45, hjust = 0.9)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S5C.cell_type_enrichment", .x), width = 3.35, height = 3.4)
)
```

### D | Developmental trajectory of each module per PsychENCODE data

```{r FigS5d.plot, fig.width = 6.5, fig.height = 3}
df_expr_mod_median %>% 
    ggplot(aes(x = window, y = median_expr, color = module, fill = module)) +
    geom_smooth(method = "loess") +
    geom_vline(xintercept = 5, linewidth = 0.2) +
    facet_wrap( ~ module, scales = "free_y", nrow = 3) +
    scale_fill_manual(values = module_colors) +
    scale_color_manual(values = module_colors) +
    labs(x = "Developmental window", y = "Median expression (PsychENCODE)", 
         title = "D | Developmental expression trajectory of each module") +
    theme(legend.position = "none",
          strip.text.x = element_text(size = 9),
          strip.background = element_rect(fill = "white"),
          axis.text.x = element_blank(),
          panel.spacing = unit(0.05, "cm"),
          strip.clip = "off"
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S5D.developmental_trajectories", .x), width = 6.5, height = 3)
)
```

# FIG S6: RCCA and GRCCA setup (generated in PowerPoint) #


# FIG S7: Gene-level GRCCA covariate weights #

```{r FigS7.plot}
df_y_res %>% 
    mutate(significant = ifelse(abs(z_score) >= 2 & p_adj < 0.05, "*", "")) %>% 
    
    ggplot(aes(x = reorder(covariate, weight), y = weight)) +
    geom_col(aes(fill = covariate, alpha = abs(weight)), color = "black") +
    geom_errorbar(aes(ymin = weight - sd, ymax = weight + sd), width = 0.2) +
    geom_text(aes(label = significant, y = 0), size = 6, vjust = 1.5) +
    geom_hline(aes(yintercept = 0), color = "black") +
    scale_fill_manual(values = c(dx_colors, rep("gray", 6))) +
    scale_alpha_continuous(range = c(0.1, 1)) +
    scale_x_discrete(labels = function(x) {str_replace(x, "_", " ")}) +
    ylim(c(-0.10, 0.10)) +
    labs(x = "", y = "weight", 
         #caption = "* = abs(z-score) >= 2 & p-adj < 0.05; \nError bar shows +/- 1 sd",
         title = "Covariate weights",
         #title = bquote(bold("A")~" | Covariate (y) coefficient weights")
    ) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 0.9))
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S7.geneLevel_covariate_weights", .x), width = fig_width, height = fig_height)
)
```


# FIG S8: Vector of significant GRCCA genes (FDR_r < 0.05 & Z-score > 2) #

```{r FigS8.plot, fig.height = 3}
df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) >= 2) %>% 
    
    ggplot(aes(x = pearsons_r, y = reorder(ensembl_gene_id, pearsons_r))) +
    geom_col(aes(fill = pearsons_r)) +
    geom_vline(xintercept = 0, color = "gray") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1, 1)) +
    guides(fill = guide_colorbar(title = expression(italic(r)[x]))) +
    #scale_color_manual(values = c("*" = "black", "-" = "transparent")) +
    xlim(c(-1, 1)) +
    labs(y = "Gene", x = expression("Structure correlation (" * italic(r)[x] * ")"), 
         title = "Significant GRCCA genes"
         #title = bquote(bold("D")~" | Significant gene structure correlations"),
    ) +
    theme(axis.text.y = element_blank(), 
          axis.ticks.y = element_blank(),
          legend.position = c(0.15, 0.70), 
          legend.box = "horizontal",
          legend.key.width = unit(0.15, "cm"),
          legend.key.height = unit(0.6, "cm")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S8.significant_GRCCA_genes_all", .x), width = fig_width, height = 3)
)
```

# FIG S9: Compare across WGCNA module solutions

### A | Model significance
```{r FigS9a, fig.width = 6.5, fig.height = 2}
df_model_res_wgcna_groups %>% 
    group_by(mod_set) %>% 
    mutate(best = ifelse(pval == min(pval), 1, 0)) %>% 
    mutate(best = ifelse(set == 5, 0, best)) %>% 
    mutate(best = factor(best)) %>%
    mutate(mod_set = str_replace(mod_set, "_", ", ")) %>% 

    ggplot(aes(x = factor(varx), y = -log10(pval))) +
    geom_hline(yintercept = -log10(0.005), color = "grey", linewidth = 0.5) +
    geom_segment(aes(y = 0, yend = -log10(pval), color = mod_set, linetype = mod_set),
                 position = position_dodge(width = 0.6)
    ) +
    geom_point(aes(fill = mod_set, size = correl, shape = best), 
               position = position_dodge(width = 0.6)
    ) +
    scale_size_continuous(range = c(1, 4)) +
    scale_shape_manual(values = c("0" = 21, "1" = 23), guide = "none") +
    scale_color_manual(values = c("actual (minSize40, cutHeight0.98)" = "#F8766D", rep("gray", 3))) +
    scale_fill_manual(values = c("actual (minSize40, cutHeight0.98)" = "#F8766D", rep("gray", 3))) +
    scale_linetype_manual(values = c("actual (minSize40, cutHeight0.98)" = "solid", 
                                     "minSize30, cutHeight0.95" = "dashed", 
                                     "minSize30, cutHeight0.98" = "dotted", 
                                     "minSize40, cutHeight0.95" = "twodash")
    ) +
    guides(fill = guide_legend(title = NULL), color = guide_legend(title = NULL),
           size = guide_legend(title = "X-Y cor"), linetype = guide_legend(title = NULL)) +
    coord_cartesian(clip = "off") +
    labs(x = "X variance explained", y = "-log10(model P value)",
         title = "A | Model significance across WGCNA solutions") +
    theme(
        legend.key.height = unit(0.2, "cm"),
        legend.title = element_text(size = 8),
        legend.position = c(0.025, 0.68),
        legend.spacing = unit(0.2, "lines")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S9A.wgcna_groups_model_significance", .x), width = 6.5, height = 2)
)
```

### B | X and Y results correlations
```{r FigS9b, fig.width = 4.0, fig.height = 3.75}
# Plot
plot_y_func <- function(data, mapping, ...) {
    ggplot(data = data, mapping = mapping) +
        geom_point(alpha = 0.5) +
        #geom_smooth(method = "lm", color = "maroon", se = FALSE, linewdith = 0.25) +
        stat_cor(aes(label = ..r.label..), size = 3, color = "maroon")
}
plot_x_func <- function(data, mapping, second_df, ...) {
    ggplot(data = second_df, mapping = mapping) +  
        geom_point(alpha = 0.5, size = 0.5) +
        stat_cor(aes(label = ..r.label..), size = 3, label.x.npc = 0.4, label.y.npc = 0.1, color = "maroon")
}

ggpairs(df_y_res_wgcna_groups %>% 
            dplyr::rename_all(~str_replace(.x, "_", ",\n")) %>% 
            dplyr::rename_at(4, ~"actual"),
        lower = list(continuous = plot_y_func),  # Scatter + best fit line
        upper = list(continuous = wrap(plot_x_func, 
                                       second_df = df_x_res_wgcna_groups %>% 
                                           dplyr::rename_all(~str_replace(.x, "_", ",\n")) %>% 
                                           dplyr::rename_at(4, ~"actual")
                                       )
        ),  # Correlation values in upper triangle
        diag = list(continuous = wrap("blank"))  # Remove density plot from diagonal
) + 
    theme_bw() +
    labs(title = "B | Model feature correlations") +
    theme(plot.title = element_text(face = "bold", size = 11),
          strip.text = element_text(size = 8.25),
          strip.clip = "off",
          strip.background = element_blank(),
          panel.spacing = unit(0.2, "lines")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S9B.wgcna_groups_x_y_cor", .x), width = 4.0, height = 3.75)
)
```

### C | Enrichment for risk genes
```{r FigS9c, fig.height = 3.75, fig.width = 2.4}
df_benchmarking_res_wgcna_groups %>%
    mutate(significant = ifelse(padj < 0.05, "yes", "no")) %>%
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), 
                          "SCZ",
                          benchmark_list),
           benchmark_list = ifelse(str_detect(benchmark_list, "SCZ"),
                                   str_remove(benchmark_list, "SCZ \\(") %>% str_remove("\\)") %>% str_replace(", ", ",\n"), 
                                   ""
           ),
           mod_set = str_replace(mod_set, "_", ",\n")
    ) %>% 
    
    ggplot(aes(x = benchmark_list, y = mod_set)) +
    geom_point(aes(fill = NES, size = -log10(padj), color = significant), 
               stroke = 1, shape = 21) +
    
    facet_wrap(vars(class), ncol = 1, scales = "free_y") +
    force_panelsizes(rows = c(1,1,1,3)) +
    scale_size_continuous(range = c(0.5, 7)) +
    scale_color_manual(values = c("yes" = "black", "no" = "transparent"), guide = "none") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1.61, 1.61)) +
    guides(fill = guide_colorbar(title = "NES", title.position = "left", title.hjust = 0.5), 
           size = guide_legend(title = "-log10(FDR)", title.position = "left", title.hjust = 0.5)) +
    coord_flip(clip = "off") +
    labs(x = NULL, y = NULL,
         title = "C | Risk gene enrichments \nacross WGCNA solutions") +
    theme(
        legend.key.height = unit(0.5, "cm"),
        legend.key.width = unit(0.25, "cm"),
        legend.title = element_text(size = 8, angle = 90, margin = margin(r = 0)),
        legend.text = element_text(margin = margin(r = -1)),
        legend.spacing = unit(0.2, "lines"),
        legend.position = c(-0.32, 0.75),
        panel.spacing = unit(0.05, "lines"),
        strip.text = element_text(size = 9),
        axis.text.x = element_text(size = 9, angle = 45, hjust = 0.9),
        plot.margin = margin(b = -10)
        #legend.margin = margin(l = -10)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S9C.wgcna_groups_benchmarking", .x), width = 2.4, height = 3.75)
)
```


# FIG S10: Compare across random module solutions

### A | Model significance
```{r FigS10a, fig.width = 6.5, fig.height = 2}
seeds <- df_model_res_random_groups %>% filter(seed != "actual") %>% pull(seed) %>% unique
df_model_res_random_groups %>% 
    group_by(seed) %>% 
    mutate(best = ifelse(pval == min(pval), 1, 0)) %>%
    mutate(best = ifelse(seed %in% c("seed1856", "seed3144", "seed3578", "seed9883") & set == 4, 0, best)) %>%
    mutate(best = factor(best)) %>%
    arrange(desc(seed)) %>% 

    ggplot(aes(x = factor(varx), y = -log10(pval))) +
    geom_hline(yintercept = -log10(0.005), color = "grey", linewidth = 0.5) +
    geom_segment(aes(y = 0, yend = -log10(pval), color = seed),
                 position = position_dodge(width = 0.80)
    ) +
    geom_point(aes(fill = seed, size = correl, shape = best), 
               position = position_dodge(width = 0.80)
    ) +
    scale_color_manual(values = c("actual" = "#F8766D", rep("gray", 10))) +
    scale_fill_manual(values = c("actual" = "#F8766D", 
                                 setNames(rep(adjustcolor("gray", alpha.f = 0.5), 10), seeds)),
                      guide = "none"
    ) +
    scale_size_continuous(range = c(1, 4)) +
    scale_shape_manual(values = c("0" = 21, "1" = 23), guide = "none") +
    guides(color = guide_legend(title = NULL), size = guide_legend(title = "X-Y cor")) +
    coord_cartesian(clip = "off") +
    labs(x = "X variance explained", y = "-log10(model P value)",
         title = "A | Model significance across random groupings") +
    theme(
        legend.key.height = unit(0.2, "cm"),
        legend.title = element_text(size = 8),
        legend.position = c(0.025, 0.60),
        legend.spacing = unit(0.5, "lines")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S10A.random_groups_model_significance", .x), width = 6.5, height = 2)
)
```

### B | X and Y results correlations
```{r FigS10b, fig.width = 6.5, fig.height = 2}
df_y_res_random_groups %>% 
    pivot_longer(contains("seed"), names_to = "seed", values_to = "random") %>%
    mutate(type = "Covariates \n(ry)") %>% 
    
    bind_rows(
        df_x_res_random_groups %>% 
            pivot_longer(contains("seed"), names_to = "seed", values_to = "random") %>% 
            mutate(type = "Genes \n(rx)")
    ) %>% 
    
    ggplot(aes(x = random, y = actual)) +
    geom_point(alpha = 0.5, size = 0.5) +
    stat_cor(aes(label = ..r.label..), label.y = 0.8, color = "maroon", size = 2.75) +
    facet_grid(type ~ seed) +
    scale_x_continuous(breaks = c(-0.5, 0.5)) +
    labs(x = "Random grouping structure correlations", y = "Actual structure correlations",
         title = "B | Model feature correlations") +
    coord_cartesian(clip = "off") +
    theme_bw() +
    theme(
        strip.text.x = element_blank(),
        strip.text.y = element_text(size = 8.5),
        strip.background = element_blank(),
        strip.clip = "off",
        plot.title = element_text(face = "bold", size = 11),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        panel.border = element_rect(linewidth = 0.30),
        plot.margin = margin(r = -5)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S10B.random_groups_x_y_cor", .x), width = 6.5, height = 2)
)
```

### C | Enrichment for risk genes
```{r FigS10c, fig.height = 2, fig.width = 6.5}
df_benchmarking_res_random_groups %>%
    mutate(significant = ifelse(padj < 0.05, "yes", "no")) %>%
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), 
                          "SCZ",
                          benchmark_list),
           benchmark_list = ifelse(str_detect(benchmark_list, "SCZ"),
                                   str_remove(benchmark_list, "SCZ \\(") %>% str_remove("\\)") %>% str_replace(", ", ",\n"), 
                                   ""
           )
    ) %>% 
    
    ggplot(aes(x = benchmark_list, y = seed)) +
    geom_point(aes(fill = NES, size = -log10(padj), color = significant), 
               stroke = 1, shape = 21) +
    
    facet_wrap(vars(class), nrow = 1, scales = "free_x") +
    force_panelsizes(cols = c(1,1,1,3)) +
    scale_size_continuous(range = c(0.1, 5)) +
    scale_color_manual(values = c("yes" = "black", "no" = "transparent"), guide = "none") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1.61, 1.61)) +
    guides(fill = guide_colorbar(title = "NES", title.position = "left", title.hjust = 0.5), 
           size = guide_legend(title = "-log10(FDR)", title.position = "left", title.hjust = 0.5)) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = NULL,
         title = "C | Risk gene enrichments across random groupings") +
    theme(
        legend.key.height = unit(0.5, "cm"),
        legend.key.width = unit(0.25, "cm"),
        legend.title = element_text(size = 8, angle = 90),
        legend.spacing = unit(0.2, "lines"),
        legend.margin = margin(l = -10)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S10C.random_groups_benchmarking", .x), width = 6.5, height = 2)
)
```


# FIG S11: Compare GRCCA and RCCA directly #

### A | Correlate GRCCA and RCCA covariate (y) structure correlations
```{r FigS11a.plot}
df_y_res_all %>% 
    pivot_wider(id_cols = covariate, names_from = analysis, values_from = pearsons_r) %>% 
    ggplot(aes(x = rcca, y = grcca)) +
    geom_point(aes(fill = rcca + grcca), shape = 21, size = 2.5) +
    geom_vline(aes(xintercept = 0), color = "gray") +
    geom_hline(aes(yintercept = 0), color = "gray") +
    geom_abline(lty = 2) +
    geom_text_repel(aes(label = covariate), size = 2.5, max.overlaps = 20, box.padding = 1.25,
                     nudge_y = 0.3, force = 150) +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1.84, 1.84), guide = "none") +
    labs(title = "A | Covariate structure correlations",
         x = "RCCA", y = "GRCCA")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S11A.covariate_structCor_correlation", .x), width = fig_width, height = fig_height)
)
```

### B | Correlate GRCCA and RCCA gene (x) structure correlations
```{r FigS11b.plot}
df_x_res_all %>% 
    pivot_wider(id_cols = ensembl_gene_id, names_from = analysis, values_from = pearsons_r) %>% 
    ggplot(aes(x = rcca, y = grcca)) +
    geom_point(aes(color = rcca + grcca), size = 1) +
    geom_vline(aes(xintercept = 0), color = "gray") +
    geom_hline(aes(yintercept = 0), color = "gray") +
    geom_abline(lty = 2) +
    stat_cor(label.sep = "\n") +
    #geom_label_repel(aes(label = gene_symbol), size = 3, max.overlaps = 50, box.padding = 1.5) +
    scale_color_gradientn(colors = gene_weight_color_scale, limits = c(-1.14, 1.14), guide = "none") +
    labs(title = "B | Gene structure correlations",
         x = "RCCA", y = "GRCCA")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S11B.gene_structCor_correlation", .x), width = fig_width, height = fig_height)
)
```

### C | Venn diagram of significant gene overlap (made in PowerPoint/Inkscape)
```{r FigS11c.prep}
## Venn diagram (overlap with Nirmala's results)
sig_grcca_genes <- df_x_res %>%
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>%
    pull(ensembl_gene_id)
sig_rcca_genes <- df_x_res_rcca %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>% 
    pull(ensembl_gene_id)
length(sig_grcca_genes) # N = 1211
length(sig_rcca_genes) # N = 2026
hypergeometric_overlap <- f_hypergeometric(sig_grcca_genes, sig_rcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 855; P < 0.0001
```

### D | RCCA results risk gene enrichment
```{r FigS11d.prep}
df_fig13d <- df_rcca_fgsea_benchmark %>% 
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), "SCZ", "other") %>% factor(levels = c("SCZ", "other"))) %>% 
    mutate(benchmark_list = str_remove_all(benchmark_list, "SCZ|\\)|\\(") %>% 
               str_trim %>% 
               str_replace(", ", ",\n")
    ) %>% 
    mutate(label_color = ifelse(padj < 0.05, "white", "black")) %>% 
    mutate(box_color = ifelse(padj < 0.05, "black", "transparent")) %>% 
    mutate(label = round(NES, 2) %>% format(digits = 3))
    #mutate(sig = ifelse(padj < 0.05, scientific(padj, digits = 3), round(padj, 2) %>% format(digits = 3)))
   # mutate(sig = case_when(padj < 0.001 ~ "***", padj < 0.01 ~ "**", padj < 0.05 ~ "*", TRUE ~ ""))
```

```{r FigS11d.plot}
df_fig13d %>% 
    ggplot(aes(x = benchmark_list, y = 1)) +
    geom_tile(aes(fill = NES, color = I(box_color)), width = 0.95, height = 0.95, linewidth = 0.75) +
    geom_text(aes(label = label, color = I(label_color)), size = 3) +
    facet_wrap(vars(class), scales = "free", nrow = 2) +
    scale_fill_gradient(low = "white", high = "#67001F", limits = c(0.838, 1.58)) +
    guides(fill = guide_colorbar(title.hjust = 0.5, title.position = "left")) +
    labs(x = NULL, y = NULL,
         title = "D | Risk gene enrichment") +
    theme(axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_blank(),
          axis.text.x = element_text(margin = margin(t = -3)),
          strip.background = element_rect(fill = "white", color = "black"),
          legend.margin = margin(r = -15),
          legend.position = "left",
          legend.justification = "top",
          legend.title = element_text(angle = 90),
          legend.key.height = unit(0.75, "cm"),
          legend.key.width = unit(0.15, "cm")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S11D.Risk_gene_enrichment", .x), width = fig_width, height = fig_height)
)
```


# FIG S12: Compare limma DEG results with Akula et al 2021 #

### A | Venn diagram of significant DEG overlap
```{r FigS12a.prep}
## Venn diagram (overlap with Nirmala's results)
my_degs <- df_limma_res %>% dplyr::filter(p_value < 0.05) %>% pull(ensembl_gene_id)
length(my_degs) # N = 1389

akula_degs <- df_akula_res %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    pull(ensembl_gene_id)
length(akula_degs) # N = 1373

sig_grcca_genes <- df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>% 
    pull(ensembl_gene_id)
length(sig_grcca_genes) # N = 1211

# Akula DEGs vs my DEGs
hypergeometric_overlap <- f_hypergeometric(my_degs, akula_degs, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 385; P = 1.9*10^-145; OR = 7.17

# GRCCA vs my DEGs
hypergeometric_overlap <- f_hypergeometric(my_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 203; P = 1.5*10^-29; OR = 2.77

# GRCCA vs Akula DEGs
hypergeometric_overlap <- f_hypergeometric(akula_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 285; P = 2.3*10^-85; OR = 5.19

# Combine the data into a named list for ggvenn
FigS12a_data <- list(
  `Current study DEGs` = my_degs,
  `Akula et al 2021 DEGs` = akula_degs,
  `Significant GRCCA genes` = sig_grcca_genes
)
```

```{r FigS12a.plot}
ggvenn(
    FigS12a_data, 
    fill_color = c("#E0E0E0", "#E0E0E0", "#E0E0E0"),
    stroke_size = 0.75,
    text_size = 3,
    set_name_size = 4,
    show_percentage = FALSE
) +
    coord_fixed(clip = "off") +
    theme_void() +
    theme(
        text = element_text(size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S12A.DEG_GRCCA_venn_diagram", .x), width = fig_width, height = fig_height)
)
```

### B | Correlation plot of L2FC
```{r FigS12b.stats}

### PERMUTE L2FC CORRELATION ###
n_perm <- 10000
df_limma_akula_cor_perm <- tibble()
set.seed(0406)
for (i in 1:n_perm) {
    
    if (i %% 100 == 0) {print(paste0("Running permutation ", i))}
    
    # Permute LFC within module
    df_tmp <- df_de_res_all %>% 
        left_join(df_modules_filt) %>%
        filter(!is.na(Akula) & metric == "l2fc") %>% 
        group_by(module) %>%
        mutate(permuted_l2fc = sample(Limma, replace = FALSE))
    
    # Calculate correlation between permuted data and Nirmala's LFC
    test <- cor.test(df_tmp$permuted_l2fc, df_tmp$Akula)
    correlation <- test$estimate
    p_value <- test$p.value

    # Append to tibble
    df_limma_akula_cor_perm <- df_limma_akula_cor_perm %>% 
        bind_rows(
            tibble(
                perm = i,
                correlation = correlation,
                p_value = p_value
            )
        )
    
}

## Calculate permutation p-values
de_limma_akula_perm_p <- df_limma_akula_cor_perm %>% 
    mutate(obs_corr = cor.test(
        df_de_res_all %>% filter(!is.na(Akula) & metric == "l2fc") %>% pull(Limma), 
        df_de_res_all %>% filter(!is.na(Akula) & metric == "l2fc") %>% pull(Akula), 
        )$estimate
    ) %>%
    summarise(
        perm_p = (sum(abs(correlation) >= abs(obs_corr)) + 1) / (n_perm + 1)
    )
#Pperm=1*10^-4


### PERMUTE T-STAT CORRELATION ###
n_perm <- 10000
df_limma_akula_cor_perm <- tibble()
set.seed(0406)
for (i in 1:n_perm) {
    
    if (i %% 100 == 0) {print(paste0("Running permutation ", i))}
    
    # Permute LFC within module
    df_tmp <- df_de_res_all %>% 
        left_join(df_modules_filt) %>%
        filter(!is.na(Akula) & metric == "t_stat") %>% 
        group_by(module) %>%
        mutate(permuted_t = sample(Limma, replace = FALSE))
    
    # Calculate correlation between permuted data and Nirmala's LFC
    test <- cor.test(df_tmp$permuted_t, df_tmp$Akula)
    correlation <- test$estimate
    p_value <- test$p.value

    # Append to tibble
    df_limma_akula_cor_perm <- df_limma_akula_cor_perm %>% 
        bind_rows(
            tibble(
                perm = i,
                correlation = correlation,
                p_value = p_value
            )
        )
    
}

## Calculate permutation p-values
de_limma_akula_perm_p <- df_limma_akula_cor_perm %>% 
    mutate(obs_corr = cor.test(
        df_de_res_all %>% filter(!is.na(Akula) & metric == "t_stat") %>% pull(Limma), 
        df_de_res_all %>% filter(!is.na(Akula) & metric == "t_stat") %>% pull(Akula), 
        )$estimate
    ) %>%
    summarise(
        perm_p = (sum(abs(correlation) >= abs(obs_corr)) + 1) / (n_perm + 1)
    )
#Pperm=1*10^-4

```

```{r FigS12b.prep}
## Create df for plot
df_figS12b <- df_de_res_all %>% 
    
    # remove genes that were not included in Nirmala's analysis
    dplyr::filter(!is.na(Akula)) %>% 
    
    # normalize values to compare across analyses
    #mutate(akula = scale(akula), current = scale(current)) %>% 
    
    # add gene labels
    mutate(label = ifelse(abs(Limma) > 1.5 & abs(Akula) > 1.5, gene_symbol, "")) %>% 
    
    # arrange to plot
    arrange(-abs(Limma))

```

```{r FigS12b.plot, fig.width = 4, fig.height = 2.25}
df_de_res_all %>% 
    #filter(metric == "l2fc") %>%
    #mutate(color = scale(Akula)[,1] + scale(Limma)[,1]) %>% 
    mutate(metric = ifelse(metric == "l2fc", "L2FC", "t-statistic")) %>% 
    
    ggplot(aes(x = Akula, y = Limma)) +
    geom_hline(yintercept = 0, color = "gray") +
    geom_vline(xintercept = 0, color = "gray") +
    geom_abline(lty = 2) +
    geom_point(color = "darkgray", size = 0.25) +
    # geom_text_repel(aes(label = label), min.segment.length = 0, max.overlaps = 15, size = 2.5) +
    geom_smooth(method = "lm", linewidth = 0.5, color = "maroon") +
    stat_cor(label.sep = "\n", size = 3, vjust = 1) +
    facet_wrap(vars(metric), scales = "free") +
    #scale_color_gradientn(colors = rev(brewer.rdbu(100)), guide = "none") +
    labs(
        x = "Akula DGE (DESeq2)", y = "Current study DGE (Limma-voom)",
        title = "B | Correlation between DGE analysis results"
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S12B.DEG_correlation", .x), width = 4, height = 2.25)
)
```


# FIG S13: Transcript model results (subset and full) #

### A | Optimal model for all transcripts
```{r FigS13a.plot}
df_results_ALLtranscripts %>% 
    mutate(significant = ifelse(pval == 0.001, "yes", "no")) %>%
    
    ggplot(aes(x = factor(varx*100), y = -log10(pval))) +
    geom_segment(aes(y = 0, yend = -log10(pval))) +
    geom_point(aes(size = correl, shape = significant, fill = -log10(pval))) +
    geom_hline(yintercept = 0, color = "darkgrey") +
    scale_shape_manual(values = c("yes" = 23, "no" = 21), guide = "none") +
    scale_fill_gradient(low = "white", high = "#67001F", limits = c(0, 3), guide = "none") +
    guides(size = guide_legend(title = "X-Y LV correlation", override.aes = list(shape = 21))) +
    coord_cartesian(clip = "off") +
    labs(x = "X variance explained (%)", y = "-log10(model p-value)",
         title = "A | Optimal variance explained \nby model (all transcripts)"
    ) +
    theme(
        legend.position = c(0.55, 0.75),
        legend.key.size = unit(0.1, "cm")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S13A.optimal_model_all_transcripts", .x), width = fig_width, height = fig_height)
)
```

### B | Covariate correlations with the latent variable (all transcripts)
```{r FigS13b.plot}
df_y_res_ALLtranscripts %>% 
    mutate(significant = ifelse(abs(z_score) >= 2 & p_adj < 0.05, "*", "")) %>% 
    
    ggplot(aes(x = pearsons_r, y = reorder(covariate, pearsons_r))) +
    geom_point(aes(size = -log10(p_adj), fill = pearsons_r), shape = 21) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    #geom_text(aes(label = significant), size = 6, vjust = 0.75, hjust = -1.5, color = "black") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-1, 1), guide = "none") +
    scale_size_continuous(range = c(2, 6)) +
    guides(size = guide_legend(title = "-log10(FDR)")) +
    coord_cartesian(clip = "off") +
    xlim(c(-1, 1)) +
    labs(y = NULL, x = expression("Structure correlation (" * italic(r)[y] * ")"), 
         title = "B | Covariate structure \ncorrelations (all transcripts)") +
    theme(legend.position = c(0.05, 0.55), 
          legend.box = "horizontal",
          #legend.justification = "center",
          legend.key.size = unit(0.1, 'cm')
          #legend.key.height = unit(0.29, "cm"),
          #legend.key.width = unit(0.30, "cm")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S13B.covariate_correlations_all_transcripts", .x), width = fig_width, height = fig_height)
)
```

### C | Optimal model for subset transcripts
```{r FigS13c.plot}
df_results_transcripts %>% 
    mutate(significant = ifelse(pval == 0.001, "yes", "no")) %>% 
    ggplot(aes(x = factor(varx*100), y = -log10(pval))) +
    geom_segment(aes(y = 0, yend = -log10(pval))) +
    geom_point(aes(size = correl, shape = significant, fill = -log10(pval))) +
    geom_hline(yintercept = 0, color = "darkgrey") +
    scale_shape_manual(values = c("yes" = 23, "no" = 21), guide = "none") +
    scale_fill_gradient(low = "white", high = "#67001F", limits = c(0, 3), guide = "none") +
    guides(size = guide_legend(title = "X-Y LV correlation", override.aes = list(shape = 21))) +
    coord_cartesian(clip = "off") +
    labs(x = "X variance explained (%)", y = "-log10(model p-value)",
         title = "C | Optimal variance explained \nby model (subset transcripts)"
    ) +
    theme(
        legend.position = c(0.55, 0.75),
        legend.key.size = unit(0.1, "cm")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S13C.optimal_model_subset_transcripts", .x), width = fig_width, height = fig_height)
)
```


# FIG S14: Isoform structures of PTK2B and ARHGAP44 (schizophrenia risk genes) #

### A | PTK2B
```{r FigS14a.plot}
## See 05.GRCCA_transcripts/03c.transcript_annotation.R for figure generation
figS14b
map(
    .x = c(".png", ".pdf"),
    .f = ~ ggsave(paste0(supplement_figures_dir, "S14A.PTK2B", .x), width = 6.5, height = 4.5)
)
```

### B | ARHGAP44
```{r FigS14a.plot}
## See 05.GRCCA_transcripts/03c.transcript_annotation.R for figure generation
figS14a
map(
    .x = c(".png", ".pdf"),
    .f = ~ ggsave(paste0(supplement_figures_dir, "S14B.ARHGAP44", .x), width = 6.5, height = 4.5)
)
```


# FIG S15: GRCCA results are related to single-cell expression gradients implicated in SCZ #

```{r FigS15.prep}
df_figS15 <- df_lf4_grcca %>% 
    left_join(df_lake_cell_type) %>% 
    dplyr::filter((type == "Astro" & cell_type == "astrocyte") |
                      (type == "Neuro-Ex" & cell_type == "glutamatergic") |
                      (type == "Neuro-In" & cell_type == "gabaergic")
    )
```

```{r FigS15.plot, fig.width = 6}
df_figS15 %>% 
    ggplot(aes(x = PEER_4, y = pearsons_r)) +
    geom_vline(aes(xintercept = 0), color = "gray") +
    geom_hline(aes(yintercept = 0), color = "gray") +
    geom_point(aes(color = pearsons_r), size = 0.5) +
    geom_smooth(color = "black", method = "lm", linewidth = 0.75) +
    stat_cor(label.sep = "\n", size = 3, 
             label.y.npc = "top", label.x.npc = "right", hjust = 1, vjust = 1) +
    facet_wrap(vars(cell_type), scales = "free") +
    scale_color_gradientn(colors = gene_weight_color_scale, guide = "none") +
    scale_x_continuous(breaks = c(-2000, 0, 2000)) +
    labs(x = "Ling et al LF4", y = expression("Structure correlation (" * italic(r)[x] * ")"),
         title = "GRCCA correlation with SNAP program by cell type")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S15.correlation_with_SNAP", .x), width = 6, height = fig_height)
)
```


# FIG S16: WGCNA module - phenotype associations determined through traditional module eigengene analysis #

### A | Module eigengene - covariate significant associations
```{r FigS16a.prep}
df_figS16a <- df_lm_dx_allPCs %>% 
    filter(module != "geneM0" & pc == "PC1") %>% 
    mutate(term = str_replace(term, "MC", "Dim")) %>% 
    mutate(term = factor(term, levels = c(paste0("Dim", 8:1), names(dx_colors)))) %>%
    mutate(significant = ifelse(p_value < 0.05, "yes", "no")) %>% 
    arrange(abs(statistic))
```

```{r FigS16a.plot, fig.height = 3, fig.width = 6.5}
df_figS16a %>% 
    ggplot(aes(x = module, y = term)) +
    geom_hline(yintercept = 8.5, color = "gray") +
    geom_point(aes(fill = statistic, size = abs(statistic), color = significant), shape = 21) +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-3.3, 3.3)) +
    scale_color_manual(values = c("yes" = "black", "no" = "grey")) +
    scale_size_continuous(limits = c(0, 3.3), range = c(0.5, 5), guide = "none") +
    guides(fill = guide_colorbar(title = "t-statistic", title.position = "top", title.hjust = 0.5),
           color = guide_legend(title = "P < 0.05", title.position = "top", title.hjust = 0.5)
    ) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = NULL,
         title = "A | Module eigengene - covariate associations"
    ) +
    theme(
        legend.position = "bottom",
        legend.justification = "center",
        legend.margin = margin(t = -10),
        legend.key.height = unit(0.15, "cm"),
        legend.key.width = unit(0.75, "cm"),
        axis.text.x = element_text(size = 8.5, angle = 45, hjust = 0.9)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S16a.module_kme_covar_matrix", .x), width = 6.5, height = 3)
)
```

### B | Modules significantly associated with SCZ per module eigengene analysis
```{r FigS16b.prep}
df_figS16b <- df_lm_dx_allPCs %>% 
    dplyr::filter(p_value < 0.05 & term == "SCZ" & pc == "PC1") %>% 
    unnest(cols = c(data)) %>% 
    mutate(label = case_when(
        dx == "SCZ" & p_value < 0.05 & p_adj > 0.05 ~ "*",
        dx == "SCZ" & p_value < 0.05 & p_adj < 0.05 ~ "**",
        TRUE ~ ""
    )
    )
```

```{r FigS16b.plot, fig.height = 3, fig.width = 6.5}
df_figS16b %>% 
    ggplot(aes(x = pc_resids, y = dx, color = dx, fill = dx)) +
    geom_point(position = position_jitter(width = 0.1), size = 0.5, alpha = 0.75) +
    geom_boxplot(alpha = 0.1, outlier.shape = NA) +
    # geom_text(aes(label = label, x = -Inf), color = "black", size = 6,
    #           vjust = 0.75, hjust = -0.1, check_overlap = TRUE) +
    facet_wrap(vars(module), ncol = 3, scales = "free_x") +
    scale_fill_manual(values = dx_colors) +
    scale_color_manual(values = dx_colors) +
    labs(x = "kME (corrected for covariates)", y = NULL,
         #caption = "* = p < 0.05; ** = FDR < 0.05",
         title = "B | Module - SCZ associations"
    ) +
    theme(legend.position = "none",
          axis.text.y = element_text(size = 10)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S16B.sig_SCZ_kME_modules", .x), width = 6.5, height = 3)
)
```


# FIG S17: WGCNA module enrichment for risk gene lists #

```{r FigS17.prep}
df_figS17 <- df_mods_risk_gene_hypergeometric %>% 
    mutate(class = ifelse(str_detect(benchmark_list, "SCZ"), "SCZ", benchmark_list) %>% factor(levels = c("SCZ", "ASD", "BD", "MDD"))) %>% 
    mutate(benchmark_list = str_remove_all(benchmark_list, "SCZ|\\)|\\(") %>% 
               str_trim %>% 
               str_replace(", ", ",\n")
    ) %>% 
    filter(module != "geneM0") %>% 
    mutate(significant = ifelse(p_adj < 0.05, "yes", "no"),
           very_significant = ifelse(p_adj < 0.001, "***", NA_character_),
           benchmark_list = ifelse(class != "SCZ", "", benchmark_list),
           module = factor(module, levels = names(rev(module_colors)))
    )
```

```{r FigS17.plot, fig.width = 5, fig.height = 5}
df_figS17 %>% 
    ggplot(aes(x = benchmark_list, y = module)) +
    # Layer 1: fill with variable alpha
    geom_point(aes(fill = module, size = -log10(p_adj), alpha = -log10(p_adj)), color = "transparent", shape = 21, stroke = 0.75) +
    # Layer 2: Border only, no alpha applied
    geom_point(aes(size = -log10(p_adj), color = significant), fill = "transparent", shape = 21, stroke = 0.75) +

    geom_text(aes(label = very_significant), size = 4, color = "black", vjust = 0.75) +
    facet_wrap(vars(class), scales = "free_x", nrow = 1) +
    force_panelsizes(cols = c(3,1,1,1)) +
    scale_alpha_continuous(guide = "none") +
    scale_fill_manual(values = module_colors, guide = "none") +
    scale_color_manual(values = c("yes" = "black", "no" = "grey"), guide = "none") +
    guides(color = guide_legend(title = "FDR < 0.05"),
           size = guide_legend(title = "-log10(FDR)")
    ) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = NULL, 
         title = "Risk gene list enrichment across WGCNA modules") +
    theme(legend.margin = margin(l = -15))
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S17.risk_gene_module_hypergeometric", .x), width = 5, height = 5)
)
```


# FIG S18: Functional enrichment of GRCCA overrepresented WGCNA modules #

### A | Hypergeometric overlap of WGCNA modules and significant GRCCA genes
```{r FigS18a.prep}
## Identify significant GRCCA genes (FDR < 0.05 & |z| > 2)
sig_grcca_genes <- df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>% 
    pull(ensembl_gene_id)
# n = 1211

## Hypergeometric test for significant overlap with modules
df_grcca_module_hypergeometric <- map_dfr(
    .x = names(module_colors),
    .f = ~ f_module_hypergeometric(
        gene_list = sig_grcca_genes, 
        gene_list_name = "", 
        wgcna_module = .x, 
        gene_universe = ensembl_gene_ids
    ) %>% 
        dplyr::select(-gene_list)
) %>% 
    mutate(p_adj = p.adjust(p_value, method = "fdr")) %>% 
    arrange(p_adj) %>% 
    mutate(module = factor(module, levels = names(module_colors))) %>% 
    
    # specify modules to label
    mutate(module_label = ifelse(p_adj < 0.05, paste0("N = ", overlap_n), ""))
```

```{r FigS18a.plot, fig.height = 5, fig.width = 2.15}
f_plot_module_hypergeometric(df_grcca_module_hypergeometric %>% mutate(p_adj = ifelse(p_adj == 0, 10^-5, p_adj)) %>% dplyr::filter(module != "geneM0"),
                             x_axis = "p-adj",
                             plot_title = "A | GRCCA significant modules",
                             #subset_x_labs = TRUE,
                             include_module_labels = TRUE,
                             module_text_size = 3,
                             legend_position = "none",
                             p_value_threshold = 0.05,
                             include_threshold_label = TRUE
)
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S18A.GRCCA_module_hypergeometric", .x), width = 2.15, height = 5)
)
```

### B | Direction of significant genes in top modules
```{r FigS18b.prep}
## Identify significant modules
sig_modules <- df_grcca_module_hypergeometric %>% dplyr::filter(module != "geneM0" & p_adj < 0.05) %>% pull(module) %>% sort()

## Identify panel sizes (number of significant genes in each module)
panel_widths <- df_x_res %>% 
    dplyr::filter(module %in% sig_modules & ensembl_gene_id %in% sig_grcca_genes) %>% 
    dplyr::count(module) %>% 
    pull(n)
```

```{r FigS18b.plot, fig.height = 5, fig.width = 2.15}
df_x_res %>% 
    dplyr::filter(module %in% sig_modules & ensembl_gene_id %in% sig_grcca_genes) %>% 
    mutate(weight = ifelse(significant == 1, weight, 0)) %>% 
    
    ggplot(aes(x = reorder_within(ensembl_gene_id, within = module, by = pearsons_r), y = pearsons_r)) +
    geom_col(aes(fill = module, alpha = abs(pearsons_r))) +
    facet_wrap(vars(module), ncol = 1, scales = "free_y") +
    force_panelsizes(rows = panel_widths) +
    scale_fill_manual(values = module_colors) +
    scale_alpha_continuous(range = c(0.3, 1)) +
    labs(title = "B | Significant module genes",
         x = "Gene", y = expression("Structure correlation (" * italic(r)[x] * ")")) +
    guides(fill = guide_legend(nrow = 2)) +
    coord_flip() +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          strip.text = element_blank(),
          legend.position = "none",
          panel.spacing = unit(0.3, "lines")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S18B.sig_module_direction", .x), width = 2.15, height = 5)
)

```

### C | GO BP enrichments of significant modules
```{r FigS18c.prep}
# Filter for GRCCA modules of interest & take top 5 pathways for each
df_figS18c <- df_mods_go %>% 
    dplyr::filter(module %in% sig_modules) %>% 
    group_by(module) %>% 
    top_n(5, wt = -log10(pvalue)) %>% 
    mutate(Description = str_wrap(Description, width = 23.5))
```

```{r FigS18c.bubbleplot, fig.height = 5, fig.width = 2.15}
df_figS18c %>% 
    ggplot(aes(x = p.adjust, y = reorder_within(Description, -p.adjust, module))) +
    geom_point(aes(size = RichFactor, fill = module), shape = 21) +
    facet_wrap(vars(module), scales = "free_y", ncol = 1) +
    scale_fill_manual(values = module_colors, guide = "none") +
    scale_y_reordered() +
    scale_x_reverse(breaks = c(0.00005, 0.00)) +
    scale_size_continuous(range = c(1, 4)) +
    coord_cartesian(clip = "off") +
    labs(x = "Enrichment FDR", y = NULL,
         title = "C | Module enrichments") +
    theme(
        legend.position = c(-1.6, 1.05),
        legend.key.size = unit(0.2, "cm"),
        axis.text = element_text(size = 8.5)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S18C.sig_GRCCA_module_enrichments", .x), width = 2.15, height = 5)
)
```


# FIG S19: Developmental expression trajectories of GRCCA genes

```{r FigS19.prep}
## Calculate the mean expression for each decile of GRCCA results across development
df_figS19 <- df_psychencode_grcca %>% #dplyr::filter(regioncode == "MFC") %>% 
    group_by(decile, window) %>% 
    summarise(median_expression = median(dev_expression),
              sd_expression = sd(dev_expression)
    )

## Rename windows with time period
psychencode_windows <- c("5-9",
                "12-13",
                "16-18",
                "19-22",
                "PCW35-PY0.3",
                "0.5-2.5",
                "2.8-10.7",
                "13-19",
                "21-64")
names(psychencode_windows) = 1:9  

## Define colors for deciles
decile_colors <- colorRampPalette(c(gene_weight_color_scale[5], "gray", gene_weight_color_scale[95]))(10) %>% rev
```

```{r FigS19, fig.height = 3, fig.width = 4}
df_figS19 %>% 
    ggplot(aes(x = window, y = median_expression, color = decile)) +
    geom_vline(xintercept = 5, color = "gray") +
    geom_smooth(method = "loess", se = FALSE) +
    scale_x_continuous(breaks = seq(1, length(psychencode_windows), 1),
                       labels = psychencode_windows) +
    scale_color_manual(values = decile_colors) +
    guides(color = guide_legend()) +
    labs(x = " PCW | PY ", y = "Median expression",
         title = "Developmental expression trends") +
    theme(
        legend.key.size = unit(0.3, "cm"),
        axis.text.x = element_text(angle = 45, hjust = 1)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S19.developmental_trajectory", .x), width = 4, height = 3)
)
```



# GRAVEYARD #


## Fig S4B alt: Module stability (alluvial)

```{r FigS4b.alluvial, fig.width = 6.5, fig.height = 4}
df_summary <- df_modules %>%
  group_by(min_size, cut_height, module, color) %>%
  summarise(n_genes = n(), .groups = "drop")

# Create the alluvial plot
library(ggalluvial)
df_summary %>% 
    mutate(
        cut_height = factor(cut_height),
        min_size = factor(min_size)
    ) %>% 
    
    ggplot(aes(x = min_size, y = n_genes, stratum = module, alluvium = module, fill = color)) +
    geom_flow(stat = "alluvium", lode.guidance = "frontback", aes(fill = color)) +
    geom_stratum(width = 0.2, color = "black") +
    scale_fill_identity() +  # Use the hex codes from the 'color' column
    facet_grid(cols = vars(cut_height)) +  # Facet into rows by min_size
    theme_minimal() +
    labs(x = "Cut Height", 
         y = "Number of Genes",
         title = "Module stability across parameters") +
    theme(
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.text = element_text(size = 10)
    )
```


## Fig SXX: Full WGCNA module functional enrichments

```{r FigS5.prep}
df_figS5 <- df_mods_go %>% 
    clean_names %>% 
    dplyr::filter(!is.na(description) & module != 0) %>% 
    
    # take top 5 paths by p-value per module to plot
    dplyr::group_by(module) %>% 
    arrange(module, -p_adjust) %>% 
    mutate(row = row_number(),
           gene_ratio = parse(text = gene_ratio) %>% eval
    ) %>% 
    top_n(n = 4, wt = row) %>% 
    #slice_max(order_by = -weight_fisher, n = 5) %>% 
    mutate(`survives FDR` = ifelse(p_adjust < 0.05, "yes", "no")) %>% 
    mutate(description = str_wrap(description, width = 20))
```

```{r FigS5.plot, fig.height = 9, fig.width = 6.5}
df_figS5 %>% 
    ggplot(aes(x = p_adjust, y = reorder_within(description, within = module, by = -p_adjust))) +
    #y = reorder_within(str_wrap(description, width = 35), -pvalue, module))) +
    geom_point(aes(size = gene_ratio, fill = ontology, color = `survives FDR`),
               shape = 21, stroke = 0.5) +
    geom_vline(aes(xintercept = 0.05), linewidth = 0.5, color = "gray") +
    facet_wrap(vars(module), scales = "free", ncol = 5) +
    scale_size_continuous(range = c(1, 3.5)) +
    scale_color_manual(values = c("yes" = "black", "no" = "transparent")) +
    scale_y_reordered() +
    scale_x_reverse(breaks = c(0.25, 0.15, 0.05, 0)) +
    coord_cartesian(clip = "off") +
    labs(y = NULL, x = "Enrichment FDR",
         title = "WGCNA module GO enrichments"
    ) +
    guides(size = guide_legend(title = "Gene ratio")) +
    theme(
        legend.position = c(0.85, 0.075),
        #legend.direction = "vertical",
        legend.title.position = "top",
        legend.title = element_text(size = 8),
        legend.key.height = unit(0.05, "cm"),
        legend.spacing.y = unit(0.1, "cm"),
        #legend.justification = "center",
        strip.background = element_rect(fill = "white"),
        strip.clip = "off",
        strip.text = element_text(size = 8),
        panel.spacing = unit(0.005, "lines"),
        axis.text.y = element_text(size = 7.5),
        axis.text.x = element_text(size = 7.0)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S5.full_GO_enrichments", .x), width = 6.5, height = 9)
)
```

Figure SXX. Full GO enrichments of gene-level modules. Top 4 Gene Ontology pathways associated with each module, determined by p-value, across all three ontologies (BP = biological process, CC = cellular component, MF = molecular function). The x-axis shows the enrichment FDR, while the y-axis shows pathway names. Each point is colored by its ontology and sized by the gene ratio (number of genes in the module in the pathway divided by the total number of genes in the pathway); points that survived correction for multiple comparisons (FDR < 0.05; vertical gray line) are outlined in black.

## Fig SXX: Module eigengene explanation

### A | What is an eigengene?
```{r FigSXXa.prep}
df_figSXXa <- df_mods_pca_covariates %>% 
    filter(module == "geneM1") %>% 
    filter(PC %in% c("PC1", "PC2")) %>% 
    mutate(PC = paste0(PC, " (", round(proportion_of_variance, 3)*100, "%)")) %>% 
    dplyr::select(sample, PC, score) %>% distinct %>% 
    pivot_wider(id_cols = sample, names_from = PC, values_from = score) %>% 
    
    mutate(dx = case_when(
        str_detect(sample, "control") ~ "Control",
        str_detect(sample, "bipolar") ~ "BD",
        str_detect(sample, "mdd") ~ "MDD",
        str_detect(sample, "schizo") ~ "SCZ"
    ) %>% 
        factor(levels = names(dx_colors))
    )
```

```{r FigSXXa.plot}
df_figSXXa %>% 
    ggplot(aes(x = `PC1 (14.9%)`,
               y = `PC2 (4.1%)`,
               fill = dx)) +
    geom_vline(aes(xintercept = 0), color = "darkgrey") +
    geom_hline(aes(yintercept = 0), color = "darkgrey") +
    geom_point(shape = 21, size = 2.0, color = "black") +
    guides(fill = guide_legend(title = NULL)) +
    scale_fill_manual(values = dx_colors) +
    labs(title = "A | Module 1 PCs") +
    theme(legend.position = "bottom",
          legend.key.width = unit(0.2, "cm"),
          legend.margin = margin(t = -15)
    )

```

### B | Covariate-module PC correlations

```{r FigSXXb.prep}
df_figSXXb <- df_lm_dx_allPCs %>% 
    filter(module != "geneM0" & proportion_of_variance >= 0.02) %>% 
    mutate(term = str_replace(term, "MC", "Dim")) %>% 
    mutate(term = factor(term, levels = c(paste0("Dim", 8:1), names(dx_colors)))) %>%
    mutate(significant = ifelse(p_adj < 0.05, "yes", "no")) %>%
    arrange(abs(statistic)) 
```


```{r FigSXXb.plot, fig.width = 6.5, fig.height = 8.5}
df_figSXXb %>% 
    ggplot(aes(x = pc, y = term)) +
    geom_hline(yintercept = 8.5, color = "darkgrey", linewidth = 0.25) +
    geom_point(aes(fill = statistic, size = abs(statistic), color = significant), shape = 21) +
    facet_wrap(vars(module), scales = "free_x") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-4.59, 4.59)) +
    scale_color_manual(values = c("yes" = "black", "no" = "grey")) +
    scale_size_continuous(limits = c(0, 4.59), range = c(0.25, 4), guide = "none") +
        guides(fill = guide_colorbar(title = "t-statistic", title.position = "top", title.hjust = 0.5),
           color = guide_legend(title = "FDR < 0.05", title.position = "top", title.hjust = 0.5)
    ) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = NULL,
         title = "Module PC-covariate associations"
    ) +
    theme(
        legend.direction = "horizontal",
        legend.box = "vertical",
        legend.position = c(0.82, 0.08),
        legend.justification = "center",
        legend.key.height = unit(0.15, "cm"),
        legend.key.width = unit(0.75, "cm"),
        axis.text.x = element_text(size = 8.0, angle = 45, hjust = 0.9),
        axis.text.y = element_text(size = 8.0),
        strip.text = element_text(size = 9)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "SXX.all_module_PC_covariate_associations", .x), width = 6.5, height = 8.5)
)
```

### C | Significant module eigengene SCZ module enrichments

```{r FigS12c.prep}
## Identify significant modules
sig_modules <- df_lm_dx_allPCs %>% 
    dplyr::filter(p_value < 0.05 & term == "SCZ" & pc == "PC1") %>% 
    pull(module)

# Filter for modules of interest & take top 5 pathways for each
df_figS12c <- df_mods_go %>% 
    dplyr::filter(module %in% sig_modules) %>% 
    group_by(module) %>% 
    top_n(5, wt = -log10(pvalue)) %>% 
    mutate(Description = str_wrap(Description, width = 30))
```

```{r FigS12c.bubbleplot, fig.height = 5, fig.width = 4.5}
df_figS12c %>% 
    ggplot(aes(x = p.adjust, y = reorder_within(Description, -p.adjust, module))) +
    geom_point(aes(size = RichFactor, fill = module), shape = 21) +
    facet_wrap(vars(module), scales = "free_y", ncol = 2) +
    scale_fill_manual(values = module_colors, guide = "none") +
    scale_y_reordered() +
    scale_x_reverse(breaks = c(0.05, 0.00)) +
    scale_size_continuous(range = c(1, 4)) +
    coord_cartesian(clip = "off") +
    labs(x = "Enrichment FDR", y = NULL,
         title = "C | SCZ module enrichments") +
    theme(legend.position = c(-1.9, 0.85))
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S12B.sig_SCZ_kME_modules", .x), width = 3.5, height = 5)
)
```




## Fig S10B alternate (word clouds)

```{r FigS10b.wordcloud, fig.height = 5}
## Identify uninteresting words
data(stop_words)
vague_terms <- c("positive", "negative", "regulation", "response", "protein", "activity",
                 "pathway", "process", "involved", "signaling")

## Identify top words in each significant (FDR < 0.05) module pathway
module_tokens_freq <- df_sig_go_res %>% 
    dplyr::filter(p.adjust < 0.05) %>% 
    group_by(module) %>% 
    unnest_tokens(word, Description) %>% 
    anti_join(stop_words) %>% 
    dplyr::filter(!(word %in% vague_terms)) %>% 
    dplyr::count(word, sort = TRUE)

## Create wordcloud plot for each module
l_figS10b <- map(
    .x = as.character(sig_modules),
    .f = ~ module_tokens_freq %>% 
        dplyr::filter(module == .x) %>% 
        top_n(10, wt = n) %>% 
        ggplot(aes(label = word, size = n, color = module)) +
        geom_text_wordcloud() +
        facet_wrap(vars(module)) +
        scale_color_manual(values = module_colors) +
        scale_size_continuous(range = c(2.5, 5)) +
        theme_void() +
        theme(strip.text = element_blank(),
              plot.margin = margin(t = -75, r = -10, b = -75, l = -10)
        )
)

## Combine and save
wrap_plots(l_figS10b, ncol = 1) & plot_annotation(title = "B | SCZ module enrichments")
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S10B.sig_SCZ_kME_modules", .x), width = fig_width, height = 5)
)
```

GO biological process pathway enrichments for each module. The top 10 words per module are shown, and the size indicates the number of times the word appears in a GO term significantly associated with that module. Words are colored by respective module.


## Fig S12C alternate (word clouds)
```{r FigS12c.prep}
## Load module GO enrichment data
#load(paste0(base_dir, "objects/", prefix, sep = "_", "SIGNED_SFT", soft_power, "_GO_RES.RDS")) # df_mods_go

## Identify words to not include
data(stop_words)
vague_terms <- c("positive", "negative", "regulation", "response", "activity",
                 "pathway", "process", "involved", "signaling")

# Filter for modules of interest
df_go_res <- df_mods_go %>% 
    dplyr::filter(module %in% sig_modules)

## Identify top words in each significant (FDR < 0.05) module pathway
module_tokens_freq <- df_go_res %>% 
    dplyr::filter(p.adjust < 0.05) %>% 
    group_by(module) %>% 
    unnest_tokens(word, Description) %>% 
    anti_join(stop_words) %>% 
    dplyr::filter(!(word %in% vague_terms)) %>% 
    dplyr::count(word, sort = TRUE)
```

```{r FigS12c.wordcloud}
l_module_wordclouds <- map(
    .x = as.character(sig_modules),
    .f = ~ module_tokens_freq %>% 
        dplyr::filter(module == .x) %>% 
        top_n(10, wt = n) %>% 
        ggplot(aes(label = word, size = n, color = module)) +
        geom_text_wordcloud() +
        facet_wrap(vars(module)) +
        scale_color_manual(values = module_colors) +
        scale_size_continuous(range = c(2.5, 5)) +
        theme_void() +
        theme(strip.text = element_blank(),
              plot.margin = margin(t = -75, r = -10, b = -75, l = -10)
        )
)
```

GO biological process pathway enrichments for each module. The top 10 words per module are shown, and the size indicates the number of times the word appears in a GO term significantly associated with that module. Words are colored by respective module.

## FIG S9: Compare current DEGs to Akula et al 2021 DEGs #

```{r FigS9.prep}
## Combine current study results with Akula 2021
df_de_akula_res <- df_de_res %>% 
    mutate(analysis = "current", .before = 1) %>% 
    
    bind_rows(df_akula_res %>% 
                  mutate(analysis = "akula") %>% 
                  dplyr::rename("gene_symbol" = "hgnc_symbol")
    )
```

### A | Venn diagram of significant DEG overlap

```{r FigS9a.prep}
## Venn diagram (overlap with Nirmala's results)
my_degs <- df_de_res %>% dplyr::filter(pvalue < 0.05) %>% pull(ensembl_gene_id)
length(my_degs) # N = 1690

akula_degs <- df_akula_res %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    pull(ensembl_gene_id)
length(akula_degs) # N = 1373

sig_grcca_genes <- df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>% 
    pull(ensembl_gene_id)
length(sig_grcca_genes) # N = 1211

# Akula DEGs vs my DEGs
hypergeometric_overlap <- f_hypergeometric(my_degs, akula_degs, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 433; P = 5.2*10^-152; OR = 6.7

# GRCCA vs my DEGs
hypergeometric_overlap <- f_hypergeometric(my_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 211; p < 0.0001; OR = 2.30

# GRCCA vs Akula DEGs
hypergeometric_overlap <- f_hypergeometric(akula_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 285; p < 0.0001; OR = 5.21

# Combine the data into a named list for ggvenn
FigS9a_data <- list(
  `Current study DEGs` = my_degs,
  `Akula et al 2021 DEGs` = akula_degs,
  `Significant GRCCA genes` = sig_grcca_genes
)
```

```{r FigS9a.plot}
ggvenn(
    FigS9a_data, 
    fill_color = c("#E0E0E0", "#E0E0E0", "#E0E0E0"),
    stroke_size = 0.75,
    text_size = 3,
    set_name_size = 4,
    show_percentage = FALSE
) +
    coord_fixed(clip = "off") +
    theme_void() +
    theme(
        text = element_text(size = 8)
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S9A.DEG_GRCCA_venn_diagram", .x), width = fig_width, height = fig_height)
)
```


### B | Correlation plot of L2FC

```{r FigS9b.prep}
## Create df for plot
df_figS9b <- df_de_akula_res %>% 
    
    # pivot wider to plot
    pivot_wider(id_cols = c(ensembl_gene_id, gene_symbol), names_from = analysis, values_from = stat) %>% 
    dplyr::filter(!is.na(akula)) %>% 
    
    # normalize values to compare across analyses
    #mutate(akula = scale(akula), current = scale(current)) %>% 
    
    # add gene labels
    mutate(label = ifelse(abs(current) > 1.5 & abs(akula) > 1.5, gene_symbol, "")) %>% 
    
    # arrange to plot
    arrange(-abs(current))

```

```{r FigS9b.stats}
n_perm <- 10000
df_de_akula_cor_perm <- tibble()
set.seed(0206)
for (i in 1:n_perm) {
    
    if (i %% 100 == 0) {print(paste0("Running permutation ", i))}
    
    # Permute LFC within module
    df_tmp <- df_figS9b %>% 
        left_join(df_modules_filt) %>% 
        group_by(module) %>% 
        mutate(permuted_lfc = sample(current, replace = FALSE))
    
    # Calculate correlation between permuted data and Nirmala's LFC
    test <- cor.test(df_tmp$permuted_lfc, df_tmp$akula)
    correlation <- test$estimate
    p_value <- test$p.value

    # Append to tibble
    df_de_akula_cor_perm <- df_de_akula_cor_perm %>% 
        bind_rows(
            tibble(
                perm = i,
                correlation = correlation,
                p_value = p_value
            )
        )
    
}

## Calculate permutation p-values
de_akula_perm_p <- df_de_akula_cor_perm %>% 
    mutate(obs_corr = cor.test(df_figS9b$current, df_figS9b$akula)$estimate) %>%
    summarise(
        perm_p = (sum(abs(correlation) >= abs(obs_corr)) + 1) / (n_perm + 1)
    )
```

```{r FigS9b.plot}
df_figS9b %>% 
    ggplot(aes(x = akula, y = current)) +
    geom_point(aes(color = akula), size = 0.75) +
    geom_hline(aes(yintercept = 0), color = "grey") +
    geom_vline(aes(xintercept = 0), color = "grey") +
    geom_smooth(method = "lm", color = "black", linewidth = 0.75) +
    geom_abline(lty = 2) +
    #geom_text(x = 2.1, y = 2.6, label = "y = x", check_overlap = TRUE) +
    geom_text_repel(aes(label = label), min.segment.length = 0, max.overlaps = 15, size = 2.5) +
    stat_cor(label.sep = "\n", vjust = 1, size = 3.5) +
    #stat_regline_equation(label.y.npc = "top", vjust = 3) +
    scale_color_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-6.2, 6.2), guide = "none") +
    #scale_color_manual(values = c("transparent", "black")) +
    #guides(color = guide_legend(title = "FDR < 0.05", title.vjust = -1.5)) +
    labs(x = "Akula DEGs t-statistic", y = "Current DEGs t-statistic", 
         title = "B | Correlation between \nDGE analysis results") +
    theme(legend.position = c(0.05, 0.7),
          legend.spacing.x = unit(0.1, "mm"),
          legend.background = element_blank(),
          legend.box.background = element_rect(colour = "black")
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "S9B.DEG_correlation", .x), width = fig_width, height = fig_height)
)
```

### C | Significant DEG overlap with significant GRCCA genes

```{r FigS9c.prep}
sig_grcca_genes <- df_x_res %>% 
    dplyr::filter(p_adj < 0.05 & abs(z_score) > 2) %>% 
    pull(ensembl_gene_id)
length(sig_grcca_genes) # N = 1211
length(my_degs) # N = 1690
length(akula_degs) # N = 1373

# GRCCA vs my DEGs
hypergeometric_overlap <- f_hypergeometric(my_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 211; p < 0.0001

# GRCCA vs Akula DEGs
hypergeometric_overlap <- f_hypergeometric(akula_degs, sig_grcca_genes, gene_universe = ensembl_gene_ids)
length(hypergeometric_overlap$overlap_genes) # N = 285; p < 0.0001

```


## Fig SXX: Cell-type enrichment of DEGs

```{r FigSXX.plot}
df_de_fgsea_cell %>% 
    mutate(sig = case_when(padj < 0.001 ~ "***", padj < 0.01 ~ "**", padj < 0.05 ~ "*", TRUE ~ "")) %>% 
    
    ggplot(aes(x = NES, y = reorder(cell_type, NES))) +
    geom_col(aes(fill = NES), color = "gray", shape = 21) +
    geom_text(aes(label = sig, x = NES), vjust = 0.75, size = 4, check_overlap = TRUE) +
    scale_fill_gradientn(colors = rev(brewer.rdbu(100)), limits = c(-2.61, 2.61)) +
    coord_cartesian(clip = "off") +
    labs(x = "NES", y = NULL,
         title = "D | Cell-type enrichment") +
    theme(legend.position = "none")
```


## Fig S13E: GRCCA vs RCCA hypergeometric overlap across structure correlation thresholds

```{r FigS13e.plot}
df_rcca_grcca_hypergeometric_benchmark %>%

    ggplot(aes(x = threshold, y = -log10(p_value))) +
    geom_hline(aes(yintercept = -log10(0.05)), color = "gray") +
    #geom_vline(aes(xintercept = length(sig_rcca_genes) / length(ensembl_ids)), lty = 2, color = "#00BFC4") +
    #geom_vline(aes(xintercept = length(sig_grcca_genes) / length(ensembl_ids)), lty = 2, color = "#F8766D") +
    geom_point(aes(color = analysis)) +
    geom_line(aes(group = analysis, color = analysis)) +
    #annotate(geom = "text", label = "GRCCA sig ", x = 0.04, y = 3.1, size = 3) +
    #annotate(geom = "text", label = "RCCA sig ", x = 0.13, y = 3.1, size = 3) +
    
    facet_wrap(vars(benchmark_list), scales = "free_y") +
    #facet_grid2(benchmark_type ~ benchmark_list, independent = "all", scales = "free") +
    labs(title = "Hypergeometric overlaps with risk gene lists across structure correlation thresholds",
         x = "Proportion of genes included (ordered by structure correlation)", y = "-log10(hypergeometric p-value)") +
    theme(legend.position = "top",
          legend.justification = "right")
```


## Figs SXX - SXX: Rare-variant related figures

### Rare variant positions in GRCCA structure correlation distribution
```{r FigSXX.GRCCA}
## Subset X res for genes in SCZ common broad list for bottom panel
df_SXX <- df_x_res_genes %>% 
    dplyr::filter(ensembl_gene_id %in% benchmarking_lists[["SCZ (rare)"]]) %>% 
    mutate(significant = ifelse(p_adj < 0.05, 1, 0) %>% factor) %>% 
    arrange(significant, abs(pearsons_r)) %>% 
    mutate(label = gene_symbol)
    #mutate(label = ifelse(ensembl_gene_id %in% top_genes, gene_symbol, ""))

## Plot aesthetics
scale_max <- df_x_res_genes %>% pull(pearsons_r) %>% abs %>% max
jitter_pos <- position_jitter(width = 0.1, seed = 123)

## Plot
figSXX <- df_x_res_genes %>%
    ggplot(aes(x = pearsons_r)) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    geom_density() +
    geom_point(data = df_SXX, 
               mapping = aes(x = pearsons_r, y = -1, fill = pearsons_r, color = significant), 
               pos = jitter_pos, shape = 21, size = 2
    ) +
    geom_text_repel(data = df_SXX,
                    mapping = aes(x = pearsons_r, y = -1, label = label),
                    pos = jitter_pos, 
                    size = 3, box.padding = 0.5, max.overlaps = 50, force = 10, min.segment.length = 0
    ) +
    
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-scale_max, scale_max), guide = "none") +
    scale_color_manual(values = c("gray", "black"), guide = "none") +
    labs(x = expression("Structure correlation (" * italic(r)[x] * ")"), y = NULL, 
         title = expression("SCZ rare variants in " * italic(r)[x] * " distribution")) +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          panel.border = element_rect(fill = "transparent", color = "black", size = 1),
          axis.line = element_blank()
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "SXX.rare_variants_in_distribution", .x), width = fig_width, height = fig_height)
)

```

### Rare variant positions in DGE distribution
```{r FigSXX.DGE}
## Subset X res for genes in SCZ common broad list for bottom panel
df_SXX_dge <- df_de_res %>% 
    dplyr::filter(ensembl_gene_id %in% benchmarking_lists[["SCZ (rare)"]]) %>% 
    mutate(significant = ifelse(pvalue < 0.05, 1, 0) %>% factor) %>% 
    arrange(significant, abs(stat)) %>% 
    mutate(label = gene_symbol)
    #mutate(label = ifelse(ensembl_gene_id %in% top_genes, gene_symbol, ""))

## Plot aesthetics
scale_max <- df_de_res %>% pull(stat) %>% abs %>% max
jitter_pos <- position_jitter(width = 0, seed = 123)

## Plot
figSXX_dge <- df_de_res %>%
    ggplot(aes(x = stat)) +
    geom_vline(xintercept = 0, color = "darkgrey") +
    geom_density() +
    geom_point(data = df_SXX_dge, 
               mapping = aes(x = stat, y = 0, fill = stat, color = significant), 
               #pos = jitter_pos, 
               shape = 21, size = 2
    ) +
    geom_text_repel(data = df_SXX_dge,
                    mapping = aes(x = stat, y = 0, label = label),
                    #pos = jitter_pos, 
                    size = 3, box.padding = 0.5, max.overlaps = 50, force = 10, min.segment.length = 0
    ) +
    
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-scale_max, scale_max), guide = "none") +
    scale_color_manual(values = c("gray", "black"), guide = "none") +
    labs(x = "DGE t-statistic", y = NULL, 
         title = "SCZ rare variants in DGE t-statistic distribution") +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          panel.border = element_rect(fill = "transparent", color = "black", size = 1),
          axis.line = element_blank()
    )
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "SXX.rare_variants_in_DGE_distribution", .x), width = fig_width, height = fig_height)
)

```

### Expression directions of rare variants

```{r FigSXX.prep}
df_figSXX <- df_vsd_regress %>% 
    dplyr::select(sample, any_of(benchmarking_lists[["SCZ (rare)"]])) %>% 
    pivot_longer(contains("ENSG"), names_to = "ensembl_gene_id", values_to = "expr") %>% 
    left_join(df_ensembl_to_symbol) %>% 
    mutate(dx = case_when(
        str_detect(sample, "control") ~ "Control",
        str_detect(sample, "bipolar") ~ "BD",
        str_detect(sample, "mdd") ~ "MDD",
        str_detect(sample, "schizo") ~ "SCZ"
    ) %>% 
        factor(levels = names(dx_colors)))
```

```{r FigSXX.plot, fig.height = 5, fig.width = 6}
figSXX <- df_figSXX %>% 
    ggplot(aes(x = dx, y = expr)) +
    geom_point(aes(color = dx), position = position_jitter(width = 0.2), size = 0.5) +
    geom_boxplot(aes(color = dx, fill = dx), alpha = 0.1, outlier.shape = NA) +
    geom_hline(yintercept = 0, color = "darkgrey") +
    facet_wrap(vars(gene_symbol), scales = "free_y", nrow = 3) +
    #force_panelsizes(cols = c(1, 1)) +
    scale_fill_manual(values = dx_colors, guide = "none") +
    scale_color_manual(values = dx_colors, guide = "none") +
    labs(x = NULL, y = "Corrected expression value",
         title = "Gene expression values of rare variants") +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9))
figSXX
map(
  .x = c(".png", ".pdf"),
  .f = ~ ggsave(paste0(supplement_figures_dir, "SXX.rare_variants_expression_directions", .x), width = 6, height = 5)
)
```

### Transcript derivatives of rare variants

```{r FigSXX}
df_x_res_transcripts %>% 
    filter(ensembl_gene_id %in% benchmarking_lists[["SCZ (rare)"]]) %>% 
    mutate(significant = ifelse(p_adj < 0.05, "yes", "no")) %>% 
    
    
    ggplot(aes(x = pearsons_r, y = reorder(transcript_symbol, pearsons_r))) +
    geom_vline(xintercept = 0, color = "gray") +
    geom_point(aes(fill = pearsons_r, size = -log10(p_adj), color = significant), shape = 21) +
    facet_wrap(vars(gene_symbol), scales = "free_y") +
    scale_fill_gradientn(colors = gene_weight_color_scale, limits = c(-0.17, 0.17), guide = "none") +
    scale_color_manual(values = c("yes" = "black", "no" = "gray"), guide = "none") +
    scale_size_continuous(range = c(1, 4)) +
    guides(size = guide_legend(title = "-log10(FDR)")) +
    coord_cartesian(clip = "off") +
    labs(x = expression("Transcript " * italic(r)[x]), y = NULL)
```

